Mirroring IT failures in `anvildev` due to 400 status from DRS

… due to "Snapshot requires an `x-user-project` header for DRS access URL requests".

<img width="2033" height="933" alt="Image" src="https://github.com/user-attachments/assets/9f44adcb-bc99-44e4-94cf-44c88282c81c" />

`[ERROR] DRSStatusException: ('Unexpected response from https://jade.datarepo-dev.broadinstitute.org/ga4gh/drs/v1/objects/v2_4ffefed7-76a2-4807-8f3b-b4dde17ad3b5/access/gcp-us-central1*74089fe0-c732-40b0-b828-d7e091f0f54b', 400, b'{""msg"":""Snapshot requires an x-user-project header for DRS access URL requests"",""status_code"":400}')`

```
Traceback (most recent call last):
  File ""/var/task/azul/chalice.py"", line 190, in patched_event_source_handler
    return old_handler(self_, event, context)
  File ""/var/task/chalice/app.py"", line 1761, in __call__
    return self.handler(event_obj)
  File ""/var/task/azul/indexer/mirror_controller.py"", line 67, in mirror
    self.mirror(event)
  File ""/var/task/azul/indexer/mirror_controller.py"", line 77, in mirror
    self._handle_events(event, action_handler)
  File ""/var/task/azul/indexer/action_controller.py"", line 50, in _handle_events
    message_handler(action)
  File ""/var/task/azul/indexer/mirror_controller.py"", line 75, in action_handler
    self.service(action.catalog).mirror(action)
  File ""/var/task/azul/indexer/mirror_service.py"", line 561, in mirror
    self._queue_actions(self._mirror(action))
  File ""/var/task/azul/indexer/mirror_service.py"", line 430, in _queue_actions
    return self._queues.send_messages(self._mirror_queue(),
  File ""/var/task/azul/queues.py"", line 210, in send_messages
    for batch in chunked(messages, self.batch_size):
  File ""/opt/python/more_itertools/recipes.py"", line 126, in take
    return list(islice(iterable, n))
  File ""/var/task/azul/indexer/mirror_service.py"", line 622, in _
    self._mirror_file(a.file)
  File ""/var/task/azul/indexer/mirror_service.py"", line 636, in _mirror_file
    file_content = self._download(file)
  File ""/var/task/azul/indexer/mirror_service.py"", line 757, in _download
    url = self._repository_url(file)
  File ""/var/task/azul/indexer/mirror_service.py"", line 752, in _repository_url
    access = object.get(AccessMethod.gs)
  File ""/var/task/azul/drs.py"", line 365, in get
    return self._get(access_method)
  File ""/var/task/azul/drs.py"", line 387, in _get
    return self._get_access(access_id, AccessMethod.https)
  File ""/var/task/azul/drs.py"", line 427, in _get_access
    raise DRSStatusException(url, response)
```

**CloudWatch Logs Insights**    
region: us-east-1    
log-group-names: /aws/lambda/azul-indexer-anvildev-mirror    
data-sources:     
facets:     
start-time: 2026-04-24T02:25:42.888Z    
end-time: 2026-04-24T02:27:36.523Z    
query-string:
  ```
  fields @timestamp, @message
| sort @timestamp asc
| limit 10000
  ```
---
| @timestamp | @message |
| --- | --- |
| 2026-04-24 02:26:26.777 | START RequestId: bdd903b0-67c2-5ef9-a8aa-61d4b13c04ed Version: 109 |
| 2026-04-24 02:26:26.780 | [INFO] 2026-04-24T02:26:26.779Z bdd903b0-67c2-5ef9-a8aa-61d4b13c04ed azul.indexer.action_controller Worker handling message SQSFifoMessage(body={&#x27;catalog&#x27;: &#x27;anvil-it&#x27;, &#x27;operation_id&#x27;: &#x27;529aa38c-b3ed-4826-a928-03a3a8387a66&#x27;, &#x27;source&#x27;: {&#x27;id&#x27;: &#x27;74089fe0-c732-40b0-b828-d7e091f0f54b&#x27;, &#x27;spec&#x27;: &#x27;tdr:bigquery:gcp:datarepo-dev-dd576076:ANVIL_CMG_Sample_1_20230225_ANV5_202512031111&#x27;, &#x27;prefix&#x27;: &#x27;&#x2f;1&#x27;, &#x27;type&#x27;: &#x27;azul.terra.TDRSourceRef&#x27;}, &#x27;prefix&#x27;: &#x27;a&#x27;, &#x27;file&#x27;: {&#x27;uuid&#x27;: &#x27;00d3b050-30ec-4eb8-81e3-2a5c2bb9513f&#x27;, &#x27;name&#x27;: &#x27;10137.merged.matefixed.sorted.markeddups.recal.bam&#x27;, &#x27;version&#x27;: &#x27;2022-06-01T00:00:00.000000Z&#x27;, &#x27;drs_uri&#x27;: &#x27;drs:&#x2f;&#x2f;jade.datarepo-dev.broadinstitute.org&#x2f;v2_4ffefed7-76a2-4807-8f3b-b4dde17ad3b5&#x27;, &#x27;size&#x27;: 1840, &#x27;content_type&#x27;: None, &#x27;md5&#x27;: &#x27;a5a2273ce9a6df96e3a0861a12a8a2ba&#x27;, &#x27;type&#x27;: &#x27;azul.plugins.metadata.anvil.AnvilFile&#x27;}, &#x27;action&#x27;: &#x27;MirrorFileAction&#x27;}, attempts=3, id=&#x27;cd432568-28c7-4d72-b64c-1d848a2b9453&#x27;, group_id=&#x27;f8a3068a-46ab-5a28-8844-892ea2007319&#x27;, dedup_id=&#x27;ff18308e-3b27-57cc-a9e8-1fb94efb416a&#x27;) |
| 2026-04-24 02:26:26.785 | [INFO] 2026-04-24T02:26:26.785Z bdd903b0-67c2-5ef9-a8aa-61d4b13c04ed azul.boto3 s3.HeadObject: Making HEAD request to https:&#x2f;&#x2f;s3.amazonaws.com&#x2f;edu-ucsc-gi-platform-anvil-dev-mirror-anvildev.us-east-1&#x2f;_it&#x2f;info&#x2f;a5a2273ce9a6df96e3a0861a12a8a2ba.md5.json |
| 2026-04-24 02:26:26.785 | [INFO] 2026-04-24T02:26:26.785Z bdd903b0-67c2-5ef9-a8aa-61d4b13c04ed azul.boto3 s3.HeadObject: … without a request body |
| 2026-04-24 02:26:26.860 | [INFO] 2026-04-24T02:26:26.860Z bdd903b0-67c2-5ef9-a8aa-61d4b13c04ed azul.boto3 s3.HeadObject: … without a response body |
| 2026-04-24 02:26:26.860 | [INFO] 2026-04-24T02:26:26.860Z bdd903b0-67c2-5ef9-a8aa-61d4b13c04ed azul.boto3 s3.HeadObject: Got 404 response |
| 2026-04-24 02:26:26.861 | [INFO] 2026-04-24T02:26:26.861Z bdd903b0-67c2-5ef9-a8aa-61d4b13c04ed azul.boto3 s3.HeadObject: Making HEAD request to https:&#x2f;&#x2f;s3.amazonaws.com&#x2f;edu-ucsc-gi-platform-anvil-dev-mirror-anvildev.us-east-1&#x2f;_it&#x2f;file&#x2f;a5a2273ce9a6df96e3a0861a12a8a2ba.md5 |
| 2026-04-24 02:26:26.861 | [INFO] 2026-04-24T02:26:26.861Z bdd903b0-67c2-5ef9-a8aa-61d4b13c04ed azul.boto3 s3.HeadObject: … without a request body |
| 2026-04-24 02:26:26.869 | [INFO] 2026-04-24T02:26:26.869Z bdd903b0-67c2-5ef9-a8aa-61d4b13c04ed azul.boto3 s3.HeadObject: Got 404 response |
| 2026-04-24 02:26:26.869 | [INFO] 2026-04-24T02:26:26.869Z bdd903b0-67c2-5ef9-a8aa-61d4b13c04ed azul.boto3 s3.HeadObject: … without a response body |
| 2026-04-24 02:26:26.869 | [INFO] 2026-04-24T02:26:26.869Z bdd903b0-67c2-5ef9-a8aa-61d4b13c04ed azul.indexer.mirror_service Mirroring file via standard upload: AnvilFile(uuid=&#x27;00d3b050-30ec-4eb8-81e3-2a5c2bb9513f&#x27;, name=&#x27;10137.merged.matefixed.sorted.markeddups.recal.bam&#x27;, version=&#x27;2022-06-01T00:00:00.000000Z&#x27;, drs_uri=&#x27;drs:&#x2f;&#x2f;jade.datarepo-dev.broadinstitute.org&#x2f;v2_4ffefed7-76a2-4807-8f3b-b4dde17ad3b5&#x27;, size=1840, content_type=None, md5=&#x27;a5a2273ce9a6df96e3a0861a12a8a2ba&#x27;) |
| 2026-04-24 02:26:26.870 | [INFO] 2026-04-24T02:26:26.870Z bdd903b0-67c2-5ef9-a8aa-61d4b13c04ed azul.terra … without a request body |
| 2026-04-24 02:26:26.870 | [INFO] 2026-04-24T02:26:26.870Z bdd903b0-67c2-5ef9-a8aa-61d4b13c04ed azul.terra Making GET request to &#x27;https:&#x2f;&#x2f;jade.datarepo-dev.broadinstitute.org&#x2f;ga4gh&#x2f;drs&#x2f;v1&#x2f;objects&#x2f;v2_4ffefed7-76a2-4807-8f3b-b4dde17ad3b5&#x27; |
| 2026-04-24 02:26:26.870 | [INFO] 2026-04-24T02:26:26.870Z bdd903b0-67c2-5ef9-a8aa-61d4b13c04ed azul.terra … with request headers [(&#x27;x-goog-api-client&#x27;, &#x27;cred-type&#x2f;sa&#x27;), (&#x27;authorization&#x27;, &#x27;REDACTED&#x27;)] |
| 2026-04-24 02:26:28.087 | [INFO] 2026-04-24T02:26:28.087Z bdd903b0-67c2-5ef9-a8aa-61d4b13c04ed azul.terra Got 200 response after 1.216s from GET to https:&#x2f;&#x2f;jade.datarepo-dev.broadinstitute.org&#x2f;ga4gh&#x2f;drs&#x2f;v1&#x2f;objects&#x2f;v2_4ffefed7-76a2-4807-8f3b-b4dde17ad3b5 |
| 2026-04-24 02:26:28.087 | [INFO] 2026-04-24T02:26:28.087Z bdd903b0-67c2-5ef9-a8aa-61d4b13c04ed azul.terra … with response headers HTTPHeaderDict({&#x27;Date&#x27;: &#x27;Fri, 24 Apr 2026 02:26:28 GMT&#x27;, &#x27;Server&#x27;: &#x27;Apache&#x27;, &#x27;X-Frame-Options&#x27;: &#x27;SAMEORIGIN&#x27;, &#x27;Access-Control-Allow-Headers&#x27;: &#x27;DNT,User-Agent,X-Requested-With,If-Modified-Since,Cache-Control,Content-Type,Range,Authorization,Accept,Referer,X-App-Id,Origin&#x27;, &#x27;Access-Control-Allow-Methods&#x27;: &#x27;GET,POST,DELETE,PUT,PATCH,OPTIONS,HEAD&#x27;, &#x27;X-Content-Type-Options&#x27;: &#x27;nosniff&#x27;, &#x27;Strict-Transport-Security&#x27;: &#x27;max-age=31536000;includeSubDomains&#x27;, &#x27;Cache-Control&#x27;: &#x27;no-cache,no-store,must-revalidate&#x27;, &#x27;X-Request-ID&#x27;: &#x27;aKl3QvXe&#x27;, &#x27;Content-Type&#x27;: &#x27;application&#x2f;json&#x27;, &#x27;Content-Length&#x27;: &#x27;2495&#x27;, &#x27;Set-Cookie&#x27;: &#x27;JSESSIONID=09458FE25A62D72F9F6CD2C8BC1155FF; Path=&#x2f;; Secure; HttpOnly; SameSite=Lax&#x27;, &#x27;Vary&#x27;: &#x27;Origin&#x27;, &#x27;Via&#x27;: &#x27;1.1 google&#x27;, &#x27;Alt-Svc&#x27;: &#x27;h3=&quot;:443&quot;; ma=2592000&#x27;}) |
| 2026-04-24 02:26:28.087 | [INFO] 2026-04-24T02:26:28.087Z bdd903b0-67c2-5ef9-a8aa-61d4b13c04ed azul.terra … with a response body of length 2495 starting in b&#x27;{&quot;id&quot;:&quot;v2_4ffefed7-76a2-4807-8f3b-b4dde17ad3b5&quot;,&quot;name&quot;:&quot;10137_10137.merged.matefixed.sorted.markeddups.recal.bam&quot;,&quot;self_uri&quot;:&quot;drs:&#x2f;&#x2f;jade.datarepo-dev.broadinstitute.org&#x2f;v2_4ffefed7-76a2-4807-8f3b-b4dde17ad3b5&quot;,&quot;size&quot;:1840,&quot;created_time&quot;:&quot;2023-02-24T20:08:34.007Z&quot;,&quot;updated_time&quot;:&quot;2023-02-24T20:08:34.007Z&quot;,&quot;version&quot;:&quot;0&quot;,&quot;mime_type&quot;:&quot;application&#x2f;octet-stream&quot;,&quot;checksums&quot;:[{&quot;checksum&quot;:&quot;c2c3a933&quot;,&quot;type&quot;:&quot;crc32c&quot;},{&quot;checksum&quot;:&quot;a5a2273ce9a6df96e3a0861a12a8a2ba&quot;,&quot;type&quot;:&quot;md5&quot;}],&quot;access_methods&quot;:[{&quot;type&quot;:&quot;gs&quot;,&quot;access_url&quot;:{&quot;url&quot;:&quot;gs:&#x2f;&#x2f;fc-73435f45-96f6-4077-8dea-89782c5c1de5&#x2f;10137&#x2f;10137.merged.matefixed.sorted.markeddups.recal.bam&quot;,&quot;headers&quot;:null},&quot;access_id&quot;:&quot;gcp-us-central1*74089fe0-c732-40b0-b828-d7e091f0f54b&quot;,&quot;region&quot;:&quot;us-central1&quot;,&quot;authorizations&quot;:{&quot;supported_types&quot;:[&quot;BearerAuth&quot;],&quot;passport_auth_issuers&quot;:null,&quot;bearer_auth_issuers&quot;:null}},{&quot;type&quot;:&quot;https&quot;,&quot;access_url&quot;:{&quot;url&quot;:&quot;https:&#x2f;&#x2f;www.googleapis.com&#x2f;storage&#x2f;v1&#x2f;b&#x2f;fc-73435f45-96f6-4077-8dea-89782c5c1de5&#x2f;o&#x2f;10137%2F10137.merged.matefixed.sorted.markeddups.recal.bam...&#x27; |
| 2026-04-24 02:26:28.088 | [INFO] 2026-04-24T02:26:28.087Z bdd903b0-67c2-5ef9-a8aa-61d4b13c04ed azul.terra Making GET request to &#x27;https:&#x2f;&#x2f;jade.datarepo-dev.broadinstitute.org&#x2f;ga4gh&#x2f;drs&#x2f;v1&#x2f;objects&#x2f;v2_4ffefed7-76a2-4807-8f3b-b4dde17ad3b5&#x2f;access&#x2f;gcp-us-central1*74089fe0-c732-40b0-b828-d7e091f0f54b&#x27; |
| 2026-04-24 02:26:28.088 | [INFO] 2026-04-24T02:26:28.088Z bdd903b0-67c2-5ef9-a8aa-61d4b13c04ed azul.terra … without a request body |
| 2026-04-24 02:26:28.088 | [INFO] 2026-04-24T02:26:28.088Z bdd903b0-67c2-5ef9-a8aa-61d4b13c04ed azul.terra … with request headers [(&#x27;x-goog-api-client&#x27;, &#x27;cred-type&#x2f;sa&#x27;), (&#x27;authorization&#x27;, &#x27;REDACTED&#x27;)] |
| 2026-04-24 02:26:29.304 | [INFO] 2026-04-24T02:26:29.304Z bdd903b0-67c2-5ef9-a8aa-61d4b13c04ed azul.terra Got 400 response after 1.216s from GET to https:&#x2f;&#x2f;jade.datarepo-dev.broadinstitute.org&#x2f;ga4gh&#x2f;drs&#x2f;v1&#x2f;objects&#x2f;v2_4ffefed7-76a2-4807-8f3b-b4dde17ad3b5&#x2f;access&#x2f;gcp-us-central1*74089fe0-c732-40b0-b828-d7e091f0f54b |
| 2026-04-24 02:26:29.304 | [INFO] 2026-04-24T02:26:29.304Z bdd903b0-67c2-5ef9-a8aa-61d4b13c04ed azul.terra … with a response body of length 98 being b&#x27;{&quot;msg&quot;:&quot;Snapshot requires an x-user-project header for DRS access URL requests&quot;,&quot;status_code&quot;:400}&#x27; |
| 2026-04-24 02:26:29.304 | [INFO] 2026-04-24T02:26:29.304Z bdd903b0-67c2-5ef9-a8aa-61d4b13c04ed azul.terra … with response headers HTTPHeaderDict({&#x27;Date&#x27;: &#x27;Fri, 24 Apr 2026 02:26:29 GMT&#x27;, &#x27;Server&#x27;: &#x27;Apache&#x27;, &#x27;X-Frame-Options&#x27;: &#x27;SAMEORIGIN&#x27;, &#x27;Access-Control-Allow-Headers&#x27;: &#x27;DNT,User-Agent,X-Requested-With,If-Modified-Since,Cache-Control,Content-Type,Range,Authorization,Accept,Referer,X-App-Id,Origin&#x27;, &#x27;Access-Control-Allow-Methods&#x27;: &#x27;GET,POST,DELETE,PUT,PATCH,OPTIONS,HEAD&#x27;, &#x27;X-Content-Type-Options&#x27;: &#x27;nosniff&#x27;, &#x27;Strict-Transport-Security&#x27;: &#x27;max-age=31536000;includeSubDomains&#x27;, &#x27;Cache-Control&#x27;: &#x27;no-cache,no-store,must-revalidate&#x27;, &#x27;X-Request-ID&#x27;: &#x27;6v5NRGMw&#x27;, &#x27;Content-Type&#x27;: &#x27;application&#x2f;json&#x27;, &#x27;Content-Length&#x27;: &#x27;98&#x27;, &#x27;Set-Cookie&#x27;: &#x27;JSESSIONID=2F4CE8C4D3C2C81301338CE62844CF44; Path=&#x2f;; Secure; HttpOnly; SameSite=Lax&#x27;, &#x27;Vary&#x27;: &#x27;Origin&#x27;, &#x27;Via&#x27;: &#x27;1.1 google&#x27;, &#x27;Alt-Svc&#x27;: &#x27;h3=&quot;:443&quot;; ma=2592000&#x27;}) |
| 2026-04-24 02:26:29.305 | [WARNING] 2026-04-24T02:26:29.304Z bdd903b0-67c2-5ef9-a8aa-61d4b13c04ed azul.indexer.action_controller Worker failed to handle message SQSFifoMessage(body={&#x27;catalog&#x27;: &#x27;anvil-it&#x27;, &#x27;operation_id&#x27;: &#x27;529aa38c-b3ed-4826-a928-03a3a8387a66&#x27;, &#x27;source&#x27;: {&#x27;id&#x27;: &#x27;74089fe0-c732-40b0-b828-d7e091f0f54b&#x27;, &#x27;spec&#x27;: &#x27;tdr:bigquery:gcp:datarepo-dev-dd576076:ANVIL_CMG_Sample_1_20230225_ANV5_202512031111&#x27;, &#x27;prefix&#x27;: &#x27;&#x2f;1&#x27;, &#x27;type&#x27;: &#x27;azul.terra.TDRSourceRef&#x27;}, &#x27;prefix&#x27;: &#x27;a&#x27;, &#x27;file&#x27;: {&#x27;uuid&#x27;: &#x27;00d3b050-30ec-4eb8-81e3-2a5c2bb9513f&#x27;, &#x27;name&#x27;: &#x27;10137.merged.matefixed.sorted.markeddups.recal.bam&#x27;, &#x27;version&#x27;: &#x27;2022-06-01T00:00:00.000000Z&#x27;, &#x27;drs_uri&#x27;: &#x27;drs:&#x2f;&#x2f;jade.datarepo-dev.broadinstitute.org&#x2f;v2_4ffefed7-76a2-4807-8f3b-b4dde17ad3b5&#x27;, &#x27;size&#x27;: 1840, &#x27;content_type&#x27;: None, &#x27;md5&#x27;: &#x27;a5a2273ce9a6df96e3a0861a12a8a2ba&#x27;, &#x27;type&#x27;: &#x27;azul.plugins.metadata.anvil.AnvilFile&#x27;}, &#x27;action&#x27;: &#x27;MirrorFileAction&#x27;}, attempts=3, id=&#x27;cd432568-28c7-4d72-b64c-1d848a2b9453&#x27;, group_id=&#x27;f8a3068a-46ab-5a28-8844-892ea2007319&#x27;, dedup_id=&#x27;ff18308e-3b27-57cc-a9e8-1fb94efb416a&#x27;) Traceback (most recent call last):   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;indexer&#x2f;action_controller.py&quot;, line 50, in _handle_events     message_handler(action)     ~~~~~~~~~~~~~~~^^^^^^^^   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;indexer&#x2f;mirror_controller.py&quot;, line 75, in action_handler     self.service(action.catalog).mirror(action)     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;indexer&#x2f;mirror_service.py&quot;, line 561, in mirror     self._queue_actions(self._mirror(action))     ~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;indexer&#x2f;mirror_service.py&quot;, line 430, in _queue_actions     return self._queues.send_messages(self._mirror_queue(),            ~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^                                       map(MirrorAction.to_sqs, actions),                                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^                                       rate_limit=rate_limit)                                       ^^^^^^^^^^^^^^^^^^^^^^   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;queues.py&quot;, line 210, in send_messages     for batch in chunked(messages, self.batch_size):                  ~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^   File &quot;&#x2f;opt&#x2f;python&#x2f;more_itertools&#x2f;recipes.py&quot;, line 126, in take     return list(islice(iterable, n))   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;indexer&#x2f;mirror_service.py&quot;, line 622, in _     self._mirror_file(a.file)     ~~~~~~~~~~~~~~~~~^^^^^^^^   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;indexer&#x2f;mirror_service.py&quot;, line 636, in _mirror_file     file_content = self._download(file)   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;indexer&#x2f;mirror_service.py&quot;, line 757, in _download     url = self._repository_url(file)   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;indexer&#x2f;mirror_service.py&quot;, line 752, in _repository_url     access = object.get(AccessMethod.gs)   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;drs.py&quot;, line 365, in get     return self._get(access_method)            ~~~~~~~~~^^^^^^^^^^^^^^^   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;drs.py&quot;, line 387, in _get     return self._get_access(access_id, AccessMethod.https)            ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;drs.py&quot;, line 427, in _get_access     raise DRSStatusException(url, response) azul.drs.DRSStatusException: (&#x27;Unexpected response from https:&#x2f;&#x2f;jade.datarepo-dev.broadinstitute.org&#x2f;ga4gh&#x2f;drs&#x2f;v1&#x2f;objects&#x2f;v2_4ffefed7-76a2-4807-8f3b-b4dde17ad3b5&#x2f;access&#x2f;gcp-us-central1*74089fe0-c732-40b0-b828-d7e091f0f54b&#x27;, 400, b&#x27;{&quot;msg&quot;:&quot;Snapshot requires an x-user-project header for DRS access URL requests&quot;,&quot;status_code&quot;:400}&#x27;) |
| 2026-04-24 02:26:29.306 | [ERROR] DRSStatusException: (&#x27;Unexpected response from https:&#x2f;&#x2f;jade.datarepo-dev.broadinstitute.org&#x2f;ga4gh&#x2f;drs&#x2f;v1&#x2f;objects&#x2f;v2_4ffefed7-76a2-4807-8f3b-b4dde17ad3b5&#x2f;access&#x2f;gcp-us-central1*74089fe0-c732-40b0-b828-d7e091f0f54b&#x27;, 400, b&#x27;{&quot;msg&quot;:&quot;Snapshot requires an x-user-project header for DRS access URL requests&quot;,&quot;status_code&quot;:400}&#x27;) Traceback (most recent call last):   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;chalice.py&quot;, line 190, in patched_event_source_handler     return old_handler(self_, event, context)   File &quot;&#x2f;var&#x2f;task&#x2f;chalice&#x2f;app.py&quot;, line 1761, in __call__     return self.handler(event_obj)   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;indexer&#x2f;mirror_controller.py&quot;, line 67, in mirror     self.mirror(event)   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;indexer&#x2f;mirror_controller.py&quot;, line 77, in mirror     self._handle_events(event, action_handler)   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;indexer&#x2f;action_controller.py&quot;, line 50, in _handle_events     message_handler(action)   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;indexer&#x2f;mirror_controller.py&quot;, line 75, in action_handler     self.service(action.catalog).mirror(action)   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;indexer&#x2f;mirror_service.py&quot;, line 561, in mirror     self._queue_actions(self._mirror(action))   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;indexer&#x2f;mirror_service.py&quot;, line 430, in _queue_actions     return self._queues.send_messages(self._mirror_queue(),   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;queues.py&quot;, line 210, in send_messages     for batch in chunked(messages, self.batch_size):   File &quot;&#x2f;opt&#x2f;python&#x2f;more_itertools&#x2f;recipes.py&quot;, line 126, in take     return list(islice(iterable, n))   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;indexer&#x2f;mirror_service.py&quot;, line 622, in _     self._mirror_file(a.file)   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;indexer&#x2f;mirror_service.py&quot;, line 636, in _mirror_file     file_content = self._download(file)   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;indexer&#x2f;mirror_service.py&quot;, line 757, in _download     url = self._repository_url(file)   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;indexer&#x2f;mirror_service.py&quot;, line 752, in _repository_url     access = object.get(AccessMethod.gs)   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;drs.py&quot;, line 365, in get     return self._get(access_method)   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;drs.py&quot;, line 387, in _get     return self._get_access(access_id, AccessMethod.https)   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;drs.py&quot;, line 427, in _get_access     raise DRSStatusException(url, response) |
| 2026-04-24 02:26:29.309 | REPORT RequestId: bdd903b0-67c2-5ef9-a8aa-61d4b13c04ed Duration: 2531.50 ms Billed Duration: 2532 ms Memory Size: 512 MB Max Memory Used: 171 MB |
| 2026-04-24 02:26:29.309 | END RequestId: bdd903b0-67c2-5ef9-a8aa-61d4b13c04ed |
---

From the most recent [IT failures in anvildev](https://gitlab.anvil.gi.ucsc.edu/ucsc/azul/-/jobs/86693), it seems this problem may also cause the `repository_files` IT test to fail via 500 response from Azul. CW logs at the bottom.

```
2026-04-24 22:15:57,554    INFO MainThread test.integration_test: Beginning sub-test [repository_files] {'catalog': 'anvil-it'}
2026-04-24 22:15:57,555    INFO MainThread test.integration_test: Making GET request to 'https://service.anvil.gi.ucsc.edu/index/files?catalog=anvil-it&filters=%7B%22files.file_format%22%3A+%7B%22is%22%3A+%5B%22.fastq%22%2C+%22.fastq.gz%22%5D%7D%7D&size=1&order=asc&sort=files.file_size'
2026-04-24 22:15:57,555    INFO MainThread test.integration_test: … without a request body
2026-04-24 22:15:57,556    INFO MainThread test.integration_test: … without request headers
2026-04-24 22:15:57,639    INFO MainThread test.integration_test: Got 200 response after 0.083s from GET to https://service.anvil.gi.ucsc.edu/index/files?catalog=anvil-it&filters=%7B%22files.file_format%22%3A+%7B%22is%22%3A+%5B%22.fastq%22%2C+%22.fastq.gz%22%5D%7D%7D&size=1&order=asc&sort=files.file_size
2026-04-24 22:15:57,639    INFO MainThread test.integration_test: … with response headers HTTPHeaderDict({'Content-Type': 'application/json', 'Content-Length': '1512', 'Connection': 'keep-alive', 'Date': 'Fri, 24 Apr 2026 22:15:57 GMT', 'X-Amzn-Trace-Id': 'Root=1-69ebeb9d-570db18e0070e8fa19abbcf5;Parent=0d85553b6ee2affd;Sampled=0;Lineage=1:87986af1:0', 'x-amzn-RequestId': '77f3a228-0f6f-44a1-b116-e3046a7207dc', 'Referrer-Policy': 'strict-origin-when-cross-origin', 'X-XSS-Protection': '1; mode=block', 'Access-Control-Allow-Origin': '*', 'Strict-Transport-Security': 'max-age=63072000; includeSubDomains; preload', 'Access-Control-Allow-Headers': 'Authorization,Content-Type,X-Amz-Date,X-Amz-Security-Token,X-Api-Key', 'X-Frame-Options': 'DENY', 'Content-Security-Policy': "default-src 'self';img-src 'self' data:;script-src 'self';style-src 'self';frame-ancestors 'none';form-action 'self'", 'x-amz-apigw-id': 'cWHArG-BIAMEayw=', 'Cache-Control': 'no-store', 'X-Content-Type-Options': 'nosniff', 'X-Cache': 'Miss from cloudfront', 'Via': '1.1 185e69fae2e25450e587a1fae1f63962.cloudfront.net (CloudFront)', 'X-Amz-Cf-Pop': 'IAD61-P3', 'X-Amz-Cf-Id': 'Vlpif7Gk3pzQjFAniGVxQ32bYYrmt8PGVUkNeSF8FPn4yu-U8YPh1w=='})
2026-04-24 22:15:57,639    INFO MainThread test.integration_test: … with a response body of length 1512 starting in b'{"hits":[],"pagination":{"count":0,"total":0,"size":1,"next":null,"previous":null,"pages":0,"sort":"files.file_size","order":"asc"},"termFacets":{"diagnoses.phenotype":{"terms":[],"total":0,"type":"terms"},"biosamples.disease":{"terms":[],"total":0,"type":"terms"},"diagnoses.disease":{"terms":[],"total":0,"type":"terms"},"diagnoses.phenopacket":{"terms":[],"total":0,"type":"terms"},"files.is_supplementary":{"terms":[],"total":0,"type":"terms"},"files.data_modality":{"terms":[],"total":0,"type":"terms"},"donors.reported_ethnicity":{"terms":[],"total":0,"type":"terms"},"datasets.consent_group":{"terms":[],"total":0,"type":"terms"},"activities.assay_type":{"terms":[],"total":0,"type":"terms"},"datasets.title":{"terms":[],"total":0,"type":"terms"},"biosamples.anatomical_site":{"terms":[],"total":0,"type":"terms"},"donors.organism_type":{"terms":[],"total":0,"type":"terms"},"datasets.data_use_permission":{"terms":[],"total":0,"type":"terms"},"donors.phenotypic_sex":{"terms":[],"total":0,"type":"terms"},"activit...'
2026-04-24 22:15:57,641    INFO MainThread test.integration_test: Making GET request to 'https://service.anvil.gi.ucsc.edu/index/files?catalog=anvil-it&filters=%7B%7D&size=1&order=asc&sort=files.file_size'
2026-04-24 22:15:57,641    INFO MainThread test.integration_test: … without a request body
2026-04-24 22:15:57,641    INFO MainThread test.integration_test: … without request headers
2026-04-24 22:15:57,699    INFO MainThread test.integration_test: Got 200 response after 0.058s from GET to https://service.anvil.gi.ucsc.edu/index/files?catalog=anvil-it&filters=%7B%7D&size=1&order=asc&sort=files.file_size
2026-04-24 22:15:57,699    INFO MainThread test.integration_test: … with response headers HTTPHeaderDict({'Content-Type': 'application/json', 'Content-Length': '4032', 'Connection': 'keep-alive', 'Date': 'Fri, 24 Apr 2026 22:15:57 GMT', 'X-Amzn-Trace-Id': 'Root=1-69ebeb9d-53650a25668e921746f345a8;Parent=79c5d56e83def8ef;Sampled=0;Lineage=1:87986af1:0', 'x-amzn-RequestId': '8d96dd5a-1b96-4454-a024-48569b6c85a9', 'Referrer-Policy': 'strict-origin-when-cross-origin', 'X-XSS-Protection': '1; mode=block', 'Access-Control-Allow-Origin': '*', 'Strict-Transport-Security': 'max-age=63072000; includeSubDomains; preload', 'Access-Control-Allow-Headers': 'Authorization,Content-Type,X-Amz-Date,X-Amz-Security-Token,X-Api-Key', 'X-Frame-Options': 'DENY', 'Content-Security-Policy': "default-src 'self';img-src 'self' data:;script-src 'self';style-src 'self';frame-ancestors 'none';form-action 'self'", 'x-amz-apigw-id': 'cWHAsE07oAMEB8A=', 'Cache-Control': 'no-store', 'X-Content-Type-Options': 'nosniff', 'X-Cache': 'Miss from cloudfront', 'Via': '1.1 185e69fae2e25450e587a1fae1f63962.cloudfront.net (CloudFront)', 'X-Amz-Cf-Pop': 'IAD61-P3', 'X-Amz-Cf-Id': '8ATcBPGbUJP-u2HSXkvdjSUR_48pqbycrb8ylVI4qurdhNUv4bc81A=='})
2026-04-24 22:15:57,699    INFO MainThread test.integration_test: … with a response body of length 4032 starting in b'{"hits":[{"entryId":"00d3b050-30ec-4eb8-81e3-2a5c2bb9513f","sources":[{"source_prefix":"/1","source_spec":"tdr:bigquery:gcp:datarepo-dev-dd576076:ANVIL_CMG_Sample_1_20230225_ANV5_202512031111","source_id":"74089fe0-c732-40b0-b828-d7e091f0f54b"}],"bundles":[{"bundle_uuid":"1bb209b9-7104-afcc-9ce9-b9a931e23d1a","bundle_version":"2022-06-01T00:00:00.000000Z"}],"activities":[{"activity_type":["Sequencing"],"assay_type":[null],"data_modality":[null]}],"biosamples":[{"anatomical_site":["sample"],"biosample_type":["sample"],"disease":[null],"donor_age_at_collection_unit":[null],"donor_age_at_collection":[{"gte":null,"lte":null}]}],"datasets":[{"dataset_id":["c70c5ba7-7571-8299-776f-b5dfba9ad4f5"],"title":["ANVIL_CMG_Sample_1"]}],"diagnoses":[],"donors":[{"organism_type":[null],"phenotypic_sex":["Female"],"reported_ethnicity":["Hispanic or Latino"],"genetic_ancestry":[null]}],"files":[{"document_id":"00d3b050-30ec-4eb8-81e3-2a5c2bb9513f","source_datarepo_row_ids":["file_inventory:d21a07bf-9ae4-4cfb-8410-f1a148e0dc...'
2026-04-24 22:15:57,701    INFO MainThread test.integration_test: Making GET request to 'https://service.anvil.gi.ucsc.edu/fetch/repository/files/00d3b050-30ec-4eb8-81e3-2a5c2bb9513f?catalog=anvil-it&version=2022-06-01T00%3A00%3A00.000000Z'
2026-04-24 22:15:57,701    INFO MainThread test.integration_test: … without a request body
2026-04-24 22:15:57,702    INFO MainThread test.integration_test: … without request headers
2026-04-24 22:16:00,504    INFO MainThread test.integration_test: Got 500 response after 2.802s from GET to https://service.anvil.gi.ucsc.edu/fetch/repository/files/00d3b050-30ec-4eb8-81e3-2a5c2bb9513f?catalog=anvil-it&version=2022-06-01T00%3A00%3A00.000000Z
2026-04-24 22:16:00,504    INFO MainThread test.integration_test: … with response headers HTTPHeaderDict({'Content-Type': 'application/json', 'Content-Length': '128', 'Connection': 'keep-alive', 'Date': 'Fri, 24 Apr 2026 22:16:00 GMT', 'X-Amzn-Trace-Id': 'Root=1-69ebeb9d-3969d1a71d4a94b645c7f145;Parent=5ac7c076ca98415a;Sampled=0;Lineage=1:87986af1:0', 'x-amzn-RequestId': '46581862-9caa-42d2-8ed3-f0d2601c82c1', 'Referrer-Policy': 'strict-origin-when-cross-origin', 'X-XSS-Protection': '1; mode=block', 'Access-Control-Allow-Origin': '*', 'Strict-Transport-Security': 'max-age=63072000; includeSubDomains; preload', 'Access-Control-Allow-Headers': 'Authorization,Content-Type,X-Amz-Date,X-Amz-Security-Token,X-Api-Key', 'X-Frame-Options': 'DENY', 'Content-Security-Policy': "default-src 'self';img-src 'self' data:;script-src 'self';style-src 'self';frame-ancestors 'none';form-action 'self'", 'x-amz-apigw-id': 'cWHAtErXoAMEgSg=', 'Cache-Control': 'no-store', 'X-Content-Type-Options': 'nosniff', 'X-Cache': 'Error from cloudfront', 'Via': '1.1 185e69fae2e25450e587a1fae1f63962.cloudfront.net (CloudFront)', 'X-Amz-Cf-Pop': 'IAD61-P3', 'X-Amz-Cf-Id': 'o_MbIhC8BeZuAcxr5vfXuFFYGf9zQ6Yd554YPa808lb3wbPomjhiHw=='})
2026-04-24 22:16:00,504    INFO MainThread test.integration_test: … with a response body of length 128 being b'{"Code":"InternalServerError","Message":"An internal server error occurred.","RequestId":"fd2dcc6a-b22b-4067-9cb7-3749258e6390"}'
2026-04-24 22:16:00,504    INFO MainThread test.integration_test: Making GET request to 'https://service.anvil.gi.ucsc.edu/fetch/repository/files/00d3b050-30ec-4eb8-81e3-2a5c2bb9513f?catalog=anvil-it&version=2022-06-01T00%3A00%3A00.000000Z'
2026-04-24 22:16:00,504    INFO MainThread test.integration_test: … without a request body
2026-04-24 22:16:00,505    INFO MainThread test.integration_test: … without request headers
2026-04-24 22:16:02,858    INFO MainThread test.integration_test: Got 500 response after 2.354s from GET to https://service.anvil.gi.ucsc.edu/fetch/repository/files/00d3b050-30ec-4eb8-81e3-2a5c2bb9513f?catalog=anvil-it&version=2022-06-01T00%3A00%3A00.000000Z
2026-04-24 22:16:02,858    INFO MainThread test.integration_test: … with response headers HTTPHeaderDict({'Content-Type': 'application/json', 'Content-Length': '128', 'Connection': 'keep-alive', 'Date': 'Fri, 24 Apr 2026 22:16:02 GMT', 'X-Amzn-Trace-Id': 'Root=1-69ebeba0-71315ba8197970cc7d9b0af5;Parent=0643f30b4cd4357f;Sampled=0;Lineage=1:87986af1:0', 'x-amzn-RequestId': 'dcd38262-d3a6-40a9-9102-9a6098eb6308', 'Referrer-Policy': 'strict-origin-when-cross-origin', 'X-XSS-Protection': '1; mode=block', 'Access-Control-Allow-Origin': '*', 'Strict-Transport-Security': 'max-age=63072000; includeSubDomains; preload', 'Access-Control-Allow-Headers': 'Authorization,Content-Type,X-Amz-Date,X-Amz-Security-Token,X-Api-Key', 'X-Frame-Options': 'DENY', 'Content-Security-Policy': "default-src 'self';img-src 'self' data:;script-src 'self';style-src 'self';frame-ancestors 'none';form-action 'self'", 'x-amz-apigw-id': 'cWHBJEmHIAMEelQ=', 'Cache-Control': 'no-store', 'X-Content-Type-Options': 'nosniff', 'X-Cache': 'Error from cloudfront', 'Via': '1.1 185e69fae2e25450e587a1fae1f63962.cloudfront.net (CloudFront)', 'X-Amz-Cf-Pop': 'IAD61-P3', 'X-Amz-Cf-Id': 'k8r2pdDZLLxj8Og94YBElRIyLCMiBQeWvfaOW7FboR5G89hZGfOrOg=='})
2026-04-24 22:16:02,858    INFO MainThread test.integration_test: … with a response body of length 128 being b'{"Code":"InternalServerError","Message":"An internal server error occurred.","RequestId":"3088362a-a930-4c2e-baf2-8faac380837e"}'
2026-04-24 22:16:02,858    INFO MainThread test.integration_test: Making GET request to 'https://service.anvil.gi.ucsc.edu/fetch/repository/files/00d3b050-30ec-4eb8-81e3-2a5c2bb9513f?catalog=anvil-it&version=2022-06-01T00%3A00%3A00.000000Z'
2026-04-24 22:16:02,858    INFO MainThread test.integration_test: … without a request body
2026-04-24 22:16:02,859    INFO MainThread test.integration_test: … without request headers
2026-04-24 22:16:04,642    INFO MainThread test.integration_test: Got 500 response after 1.783s from GET to https://service.anvil.gi.ucsc.edu/fetch/repository/files/00d3b050-30ec-4eb8-81e3-2a5c2bb9513f?catalog=anvil-it&version=2022-06-01T00%3A00%3A00.000000Z
2026-04-24 22:16:04,642    INFO MainThread test.integration_test: … with response headers HTTPHeaderDict({'Content-Type': 'application/json', 'Content-Length': '128', 'Connection': 'keep-alive', 'Date': 'Fri, 24 Apr 2026 22:16:04 GMT', 'X-Amzn-Trace-Id': 'Root=1-69ebeba2-3d69ec490a88ad8a51eeb8ca;Parent=787cafe95f4b1879;Sampled=0;Lineage=1:87986af1:0', 'x-amzn-RequestId': '745031df-9002-4e41-8212-3d9462beeaa7', 'Referrer-Policy': 'strict-origin-when-cross-origin', 'X-XSS-Protection': '1; mode=block', 'Access-Control-Allow-Origin': '*', 'Strict-Transport-Security': 'max-age=63072000; includeSubDomains; preload', 'Access-Control-Allow-Headers': 'Authorization,Content-Type,X-Amz-Date,X-Amz-Security-Token,X-Api-Key', 'X-Frame-Options': 'DENY', 'Content-Security-Policy': "default-src 'self';img-src 'self' data:;script-src 'self';style-src 'self';frame-ancestors 'none';form-action 'self'", 'x-amz-apigw-id': 'cWHBgHQ-IAMEncw=', 'Cache-Control': 'no-store', 'X-Content-Type-Options': 'nosniff', 'X-Cache': 'Error from cloudfront', 'Via': '1.1 185e69fae2e25450e587a1fae1f63962.cloudfront.net (CloudFront)', 'X-Amz-Cf-Pop': 'IAD61-P3', 'X-Amz-Cf-Id': '-NdYdMDqASNSBLJha8ieiIpK9NA8lTh8X4x7lMRus-xr2swwsMOiTg=='})
2026-04-24 22:16:04,642    INFO MainThread test.integration_test: … with a response body of length 128 being b'{"Code":"InternalServerError","Message":"An internal server error occurred.","RequestId":"607bb778-223c-48f0-b0c9-5ec77511c90d"}'
2026-04-24 22:16:04,642    INFO MainThread test.integration_test: Making GET request to 'https://service.anvil.gi.ucsc.edu/fetch/repository/files/00d3b050-30ec-4eb8-81e3-2a5c2bb9513f?catalog=anvil-it&version=2022-06-01T00%3A00%3A00.000000Z'
2026-04-24 22:16:04,642    INFO MainThread test.integration_test: … without a request body
2026-04-24 22:16:04,643    INFO MainThread test.integration_test: … without request headers
2026-04-24 22:16:06,065    INFO MainThread test.integration_test: Got 500 response after 1.422s from GET to https://service.anvil.gi.ucsc.edu/fetch/repository/files/00d3b050-30ec-4eb8-81e3-2a5c2bb9513f?catalog=anvil-it&version=2022-06-01T00%3A00%3A00.000000Z
2026-04-24 22:16:06,073    INFO MainThread test.integration_test: … with response headers HTTPHeaderDict({'Content-Type': 'application/json', 'Content-Length': '128', 'Connection': 'keep-alive', 'Date': 'Fri, 24 Apr 2026 22:16:06 GMT', 'X-Amzn-Trace-Id': 'Root=1-69ebeba4-7c3214181fea73bb381b889b;Parent=64e71e4033b23e37;Sampled=0;Lineage=1:87986af1:0', 'x-amzn-RequestId': '26dc8d4c-db66-450f-a33d-24eee2a3c170', 'Referrer-Policy': 'strict-origin-when-cross-origin', 'X-XSS-Protection': '1; mode=block', 'Access-Control-Allow-Origin': '*', 'Strict-Transport-Security': 'max-age=63072000; includeSubDomains; preload', 'Access-Control-Allow-Headers': 'Authorization,Content-Type,X-Amz-Date,X-Amz-Security-Token,X-Api-Key', 'X-Frame-Options': 'DENY', 'Content-Security-Policy': "default-src 'self';img-src 'self' data:;script-src 'self';style-src 'self';frame-ancestors 'none';form-action 'self'", 'x-amz-apigw-id': 'cWHByEw-oAMEtFA=', 'Cache-Control': 'no-store', 'X-Content-Type-Options': 'nosniff', 'X-Cache': 'Error from cloudfront', 'Via': '1.1 185e69fae2e25450e587a1fae1f63962.cloudfront.net (CloudFront)', 'X-Amz-Cf-Pop': 'IAD61-P3', 'X-Amz-Cf-Id': 'yBHgsIqWastIerplSogPvh3o9p3dWvWDLSX-5DS6-5CHCqjxg-Jyyw=='})
2026-04-24 22:16:06,073    INFO MainThread test.integration_test: … with a response body of length 128 being b'{"Code":"InternalServerError","Message":"An internal server error occurred.","RequestId":"4addce63-5aa3-4b90-800e-901340eb8030"}'
2026-04-24 22:16:06,073    INFO MainThread test.integration_test: Making GET request to 'https://service.anvil.gi.ucsc.edu/fetch/repository/files/00d3b050-30ec-4eb8-81e3-2a5c2bb9513f?catalog=anvil-it&version=2022-06-01T00%3A00%3A00.000000Z'
2026-04-24 22:16:06,073    INFO MainThread test.integration_test: … without a request body
2026-04-24 22:16:06,074    INFO MainThread test.integration_test: … without request headers
2026-04-24 22:16:07,437    INFO MainThread test.integration_test: Got 500 response after 1.364s from GET to https://service.anvil.gi.ucsc.edu/fetch/repository/files/00d3b050-30ec-4eb8-81e3-2a5c2bb9513f?catalog=anvil-it&version=2022-06-01T00%3A00%3A00.000000Z
2026-04-24 22:16:07,437    INFO MainThread test.integration_test: … with response headers HTTPHeaderDict({'Content-Type': 'application/json', 'Content-Length': '128', 'Connection': 'keep-alive', 'Date': 'Fri, 24 Apr 2026 22:16:07 GMT', 'X-Amzn-Trace-Id': 'Root=1-69ebeba6-4fd55a3d4b74f41169833747;Parent=5ca8f7adb4e5e904;Sampled=0;Lineage=1:87986af1:0', 'x-amzn-RequestId': 'f3613313-1509-4ebf-a901-7274da605558', 'Referrer-Policy': 'strict-origin-when-cross-origin', 'X-XSS-Protection': '1; mode=block', 'Access-Control-Allow-Origin': '*', 'Strict-Transport-Security': 'max-age=63072000; includeSubDomains; preload', 'Access-Control-Allow-Headers': 'Authorization,Content-Type,X-Amz-Date,X-Amz-Security-Token,X-Api-Key', 'X-Frame-Options': 'DENY', 'Content-Security-Policy': "default-src 'self';img-src 'self' data:;script-src 'self';style-src 'self';frame-ancestors 'none';form-action 'self'", 'x-amz-apigw-id': 'cWHCAFGNoAMEvZA=', 'Cache-Control': 'no-store', 'X-Content-Type-Options': 'nosniff', 'X-Cache': 'Error from cloudfront', 'Via': '1.1 185e69fae2e25450e587a1fae1f63962.cloudfront.net (CloudFront)', 'X-Amz-Cf-Pop': 'IAD61-P3', 'X-Amz-Cf-Id': 'U53oaaodG3dj1gwK5K6FRxq1gsvJ2XLW_l9fYL9mtQGGzlRaj_zjNg=='})
2026-04-24 22:16:07,438    INFO MainThread test.integration_test: … with a response body of length 128 being b'{"Code":"InternalServerError","Message":"An internal server error occurred.","RequestId":"6bf40948-42ec-4231-95ff-0dfb140ee877"}'
2026-04-24 22:16:07,438    INFO MainThread test.integration_test: Making GET request to 'https://service.anvil.gi.ucsc.edu/fetch/repository/files/00d3b050-30ec-4eb8-81e3-2a5c2bb9513f?catalog=anvil-it&version=2022-06-01T00%3A00%3A00.000000Z'
2026-04-24 22:16:07,438    INFO MainThread test.integration_test: … without a request body
2026-04-24 22:16:07,440    INFO MainThread test.integration_test: … without request headers
2026-04-24 22:16:08,672    INFO MainThread test.integration_test: Got 500 response after 1.233s from GET to https://service.anvil.gi.ucsc.edu/fetch/repository/files/00d3b050-30ec-4eb8-81e3-2a5c2bb9513f?catalog=anvil-it&version=2022-06-01T00%3A00%3A00.000000Z
2026-04-24 22:16:08,672    INFO MainThread test.integration_test: … with response headers HTTPHeaderDict({'Content-Type': 'application/json', 'Content-Length': '128', 'Connection': 'keep-alive', 'Date': 'Fri, 24 Apr 2026 22:16:08 GMT', 'X-Amzn-Trace-Id': 'Root=1-69ebeba7-209781c148e8d26758344741;Parent=02c0d679b83c5fe2;Sampled=0;Lineage=1:87986af1:0', 'x-amzn-RequestId': '1fbb5dda-4044-4e98-9f43-1e0a87fa5500', 'Referrer-Policy': 'strict-origin-when-cross-origin', 'X-XSS-Protection': '1; mode=block', 'Access-Control-Allow-Origin': '*', 'Strict-Transport-Security': 'max-age=63072000; includeSubDomains; preload', 'Access-Control-Allow-Headers': 'Authorization,Content-Type,X-Amz-Date,X-Amz-Security-Token,X-Api-Key', 'X-Frame-Options': 'DENY', 'Content-Security-Policy': "default-src 'self';img-src 'self' data:;script-src 'self';style-src 'self';frame-ancestors 'none';form-action 'self'", 'x-amz-apigw-id': 'cWHCOF9UIAMEvzQ=', 'Cache-Control': 'no-store', 'X-Content-Type-Options': 'nosniff', 'X-Cache': 'Error from cloudfront', 'Via': '1.1 185e69fae2e25450e587a1fae1f63962.cloudfront.net (CloudFront)', 'X-Amz-Cf-Pop': 'IAD61-P3', 'X-Amz-Cf-Id': 'YiD0Rss-m12N_rnkOu1XwR7nfI4-00NewLvRn9LX9n-p9yIVkILk0g=='})
2026-04-24 22:16:08,672    INFO MainThread test.integration_test: … with a response body of length 128 being b'{"Code":"InternalServerError","Message":"An internal server error occurred.","RequestId":"393b6012-d7d9-4596-8661-241491e84d11"}'
2026-04-24 22:16:08,672    INFO MainThread test.integration_test: Failed sub-test [repository_files] {'catalog': 'anvil-it'}
  test_indexing (integration_test.IndexingIntegrationTest.test_indexing) [repository_files] (catalog='anvil-it') ... ERROR
…
======================================================================
ERROR: test_indexing (integration_test.IndexingIntegrationTest.test_indexing) [repository_files] (catalog='anvil-it')
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builds/ucsc/azul/test/integration_test.py", line 402, in subTest
    yield
  File "/builds/ucsc/azul/test/integration_test.py", line 1189, in _test_repository_files
    self._test_file_download(source, inner_file)
    ~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^
  File "/builds/ucsc/azul/test/integration_test.py", line 1201, in _test_file_download
    response = self._get_url_unchecked(GET, file_url)
  File "/builds/ucsc/azul/test/integration_test.py", line 1014, in _get_url_unchecked
    response = http.request(method=method,
                            url=url,
    ...<2 lines>...
                            timeout=float(config.api_gateway_lambda_timeout + 1),
                            preload_content=not stream)
  File "/build/.venv/lib/python3.14/site-packages/urllib3/_request_methods.py", line 135, in request
    return self.request_encode_url(
           ~~~~~~~~~~~~~~~~~~~~~~~^
        method,
        ^^^^^^^
    ...<3 lines>...
        **urlopen_kw,
        ^^^^^^^^^^^^^
    )
    ^
  File "/build/.venv/lib/python3.14/site-packages/urllib3/_request_methods.py", line 182, in request_encode_url
    return self.urlopen(method, url, **extra_kw)
           ~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/builds/ucsc/azul/src/azul/http.py", line 446, in urlopen
    raise urllib3.exceptions.MaxRetryError(pool, url)
urllib3.exceptions.MaxRetryError: test.integration_test.LoggingHTTPSConnectionPool(host='service.anvil.gi.ucsc.edu', port=443): Max retries exceeded with url: https://service.anvil.gi.ucsc.edu/fetch/repository/files/00d3b050-30ec-4eb8-81e3-2a5c2bb9513f?catalog=anvil-it&version=2022-06-01T00%3A00%3A00.000000Z (Caused by None)
----------------------------------------------------------------------
Ran 16 tests in 392.103s
FAILED (errors=1)
make: *** [Makefile:288: integration_test] Error 1
Cleaning up project directory and file based variables 00:01
ERROR: Job failed: exit code 1
```

Azul error traceback:
```

[ERROR]	2026-04-24T22:16:08.662Z	393b6012-d7d9-4596-8661-241491e84d11	azul-service-anvildev	Caught exception for path /fetch/repository/files/{file_uuid}
Traceback (most recent call last):
  File ""/var/task/chalice/app.py"", line 1922, in _get_view_function_response
    response = view_function(**function_args)
  File ""/var/task/azul/service/repository_controller.py"", line 261, in get_fetch_repository_files
    body = self.download_file(file_uuid, fetch=True)
  File ""/var/task/azul/service/repository_controller.py"", line 313, in download_file
    return self._download_file(catalog=catalog,
           ~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^
                               fetch=fetch,
                               ^^^^^^^^^^^^
    ...<2 lines>...
                               headers=headers,
                               ^^^^^^^^^^^^^^^^
                               authentication=authentication)
                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File ""/var/task/azul/service/repository_controller.py"", line 417, in _download_file
    download.update(authentication)
    ~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^
  File ""/var/task/azul/plugins/repository/tdr.py"", line 285, in update
    access = drs_client.get(access_method=AccessMethod.gs)
  File ""/var/task/azul/drs.py"", line 365, in get
    return self._get(access_method)
           ~~~~~~~~~^^^^^^^^^^^^^^^
  File ""/var/task/azul/drs.py"", line 387, in _get
    return self._get_access(access_id, AccessMethod.https)
           ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File ""/var/task/azul/drs.py"", line 427, in _get_access
    raise DRSStatusException(url, response)
azul.drs.DRSStatusException: ('Unexpected response from https://jade.datarepo-dev.broadinstitute.org/ga4gh/drs/v1/objects/v2_4ffefed7-76a2-4807-8f3b-b4dde17ad3b5/access/gcp-us-central1*74089fe0-c732-40b0-b828-d7e091f0f54b', 400, b'{""msg"":""Snapshot requires an x-user-project header for DRS access URL requests"",""status_code"":400}')"
```


**CloudWatch Logs Insights**    
region: us-east-1    
log-group-names: /aws/lambda/azul-service-anvildev    
data-sources:     
facets:     
start-time: -86400s    
end-time: 0s    
query-string:
  ```
  fields @timestamp, @message
# | filter @message like /Returning 5|ServiceUnavailableError/
# | filter @message like "fetch/repository/files/00d3b050-30ec-4eb8-81e3-2a5c2bb9513f"
| filter @message like /393b6012/
| sort @timestamp asc
| limit 1000
  ```
---
| @timestamp | @message |
| --- | --- |
| 2026-04-24 22:16:07.471 | START RequestId: 393b6012-d7d9-4596-8661-241491e84d11 Version: 138 |
| 2026-04-24 22:16:07.472 | [INFO] 2026-04-24T22:16:07.472Z 393b6012-d7d9-4596-8661-241491e84d11 azul.chalice Received GET request for &#x27;&#x2f;fetch&#x2f;repository&#x2f;files&#x2f;00d3b050-30ec-4eb8-81e3-2a5c2bb9513f&#x27;, with {&quot;query&quot;: {&quot;catalog&quot;: &quot;anvil-it&quot;, &quot;version&quot;: &quot;2022-06-01T00:00:00.000000Z&quot;}, &quot;headers&quot;: {&quot;accept-encoding&quot;: &quot;identity&quot;, &quot;cloudfront-forwarded-proto&quot;: &quot;https&quot;, &quot;cloudfront-is-desktop-viewer&quot;: &quot;true&quot;, &quot;cloudfront-is-mobile-viewer&quot;: &quot;false&quot;, &quot;cloudfront-is-smarttv-viewer&quot;: &quot;false&quot;, &quot;cloudfront-is-tablet-viewer&quot;: &quot;false&quot;, &quot;cloudfront-viewer-asn&quot;: &quot;14618&quot;, &quot;cloudfront-viewer-country&quot;: &quot;US&quot;, &quot;host&quot;: &quot;service.anvil.gi.ucsc.edu&quot;, &quot;user-agent&quot;: &quot;python-urllib3&#x2f;2.6.3&quot;, &quot;via&quot;: &quot;1.1 185e69fae2e25450e587a1fae1f63962.cloudfront.net (CloudFront)&quot;, &quot;x-amz-cf-id&quot;: &quot;YiD0Rss-m12N_rnkOu1XwR7nfI4-00NewLvRn9LX9n-p9yIVkILk0g==&quot;, &quot;x-amzn-trace-id&quot;: &quot;Root=1-69ebeba7-209781c148e8d26758344741&quot;, &quot;x-forwarded-for&quot;: &quot;54.211.146.213, 18.68.30.172&quot;, &quot;x-forwarded-port&quot;: &quot;443&quot;, &quot;x-forwarded-proto&quot;: &quot;https&quot;}}. |
| 2026-04-24 22:16:07.472 | [INFO] 2026-04-24T22:16:07.472Z 393b6012-d7d9-4596-8661-241491e84d11 azul.chalice Did not authenticate request. |
| 2026-04-24 22:16:07.472 | [INFO] 2026-04-24T22:16:07.472Z 393b6012-d7d9-4596-8661-241491e84d11 azul.chalice … without a request body |
| 2026-04-24 22:16:07.473 | [INFO] 2026-04-24T22:16:07.473Z 393b6012-d7d9-4596-8661-241491e84d11 opensearch … with a request body of length 399 being b&#x27;{&quot;query&quot;:{&quot;bool&quot;:{&quot;must&quot;:[{&quot;constant_score&quot;:{&quot;filter&quot;:{&quot;terms&quot;:{&quot;contents.files.document_id.keyword&quot;:[&quot;00d3b050-30ec-4eb8-81e3-2a5c2bb9513f&quot;]}}}},{&quot;constant_score&quot;:{&quot;filter&quot;:{&quot;terms&quot;:{&quot;contents.files.version.keyword&quot;:[&quot;2022-06-01T00:00:00.000000Z&quot;]}}}},{&quot;constant_score&quot;:{&quot;filter&quot;:{&quot;terms&quot;:{&quot;sources.id.keyword&quot;:[&quot;74089fe0-c732-40b0-b828-d7e091f0f54b&quot;,&quot;790795c4-49b1-4ac8-a060-207b92ea08c5&quot;]}}}}]}}}&#x27; |
| 2026-04-24 22:16:07.473 | [INFO] 2026-04-24T22:16:07.473Z 393b6012-d7d9-4596-8661-241491e84d11 opensearch Making POST request to https:&#x2f;&#x2f;vpc-azul-index-anvildev-jubbechxp4wxzfj2kf2hiewjgi.us-east-1.es.amazonaws.com:443&#x2f;azul_v2_anvildev_anvil-it_files_aggregate&#x2f;_search |
| 2026-04-24 22:16:07.483 | [INFO] 2026-04-24T22:16:07.483Z 393b6012-d7d9-4596-8661-241491e84d11 opensearch Got 200 response after 0.010s from POST to https:&#x2f;&#x2f;vpc-azul-index-anvildev-jubbechxp4wxzfj2kf2hiewjgi.us-east-1.es.amazonaws.com:443&#x2f;azul_v2_anvildev_anvil-it_files_aggregate&#x2f;_search |
| 2026-04-24 22:16:07.483 | [INFO] 2026-04-24T22:16:07.483Z 393b6012-d7d9-4596-8661-241491e84d11 opensearch … with a response body of length 2762 starting in &#x27;{&quot;took&quot;:2,&quot;timed_out&quot;:false,&quot;_shards&quot;:{&quot;total&quot;:1,&quot;successful&quot;:1,&quot;skipped&quot;:0,&quot;failed&quot;:0},&quot;hits&quot;:{&quot;total&quot;:{&quot;value&quot;:1,&quot;relation&quot;:&quot;eq&quot;},&quot;max_score&quot;:3.0,&quot;hits&quot;:[{&quot;_index&quot;:&quot;azul_v2_anvildev_anvil-it_files_aggregate&quot;,&quot;_id&quot;:&quot;00d3b050-30ec-4eb8-81e3-2a5c2bb9513f&quot;,&quot;_score&quot;:3.0,&quot;_source&quot;:{&quot;entity_id&quot;:&quot;00d3b050-30ec-4eb8-81e3-2a5c2bb9513f&quot;,&quot;contents&quot;:{&quot;activities&quot;:[{&quot;document_id&quot;:[&quot;e0cd75c0-ab72-4456-a8db-47e54988191e&quot;],&quot;source_datarepo_row_ids&quot;:[&quot;sequencing:40a6eff9-3886-49ce-b3c0-0002ef8fef2d&quot;],&quot;activity_id&quot;:[&quot;87ad00ec-986e-bf1c-c139-f755d601a669&quot;],&quot;activity_table&quot;:[&quot;anvil_sequencingactivity&quot;],&quot;activity_type&quot;:[&quot;Sequencing&quot;],&quot;assay_type&quot;:[&quot;~null&quot;],&quot;data_modality&quot;:[&quot;~null&quot;],&quot;reference_assembly&quot;:[&quot;~null&quot;]}],&quot;biosamples&quot;:[{&quot;document_id&quot;:[&quot;1bb209b9-7104-4fcc-9ce9-b9a931e23d1a&quot;],&quot;source_datarepo_row_ids&quot;:[&quot;sample:badc7520-3f9e-4f6e-9397-85c9b4082406&quot;],&quot;biosample_id&quot;:[&quot;715e8ba4-6f8b-c29c-24f3-fbfdb8d96d71&quot;],&quot;anatomical_site&quot;:[&quot;sample&quot;],&quot;apriori_cell_type&quot;:[&quot;~null&quot;],&quot;biosample_type&quot;:[&quot;sample&quot;],&quot;disease&quot;:[&quot;~null&quot;],&quot;donor_age_a…&#x27; |
| 2026-04-24 22:16:07.485 | [INFO] 2026-04-24T22:16:07.485Z 393b6012-d7d9-4596-8661-241491e84d11 azul.boto3 s3.HeadObject: Making HEAD request to https:&#x2f;&#x2f;s3.amazonaws.com&#x2f;edu-ucsc-gi-platform-anvil-dev-mirror-anvildev.us-east-1&#x2f;_it&#x2f;info&#x2f;a5a2273ce9a6df96e3a0861a12a8a2ba.md5.json |
| 2026-04-24 22:16:07.485 | [INFO] 2026-04-24T22:16:07.485Z 393b6012-d7d9-4596-8661-241491e84d11 azul.boto3 s3.HeadObject: … without a request body |
| 2026-04-24 22:16:07.499 | [INFO] 2026-04-24T22:16:07.499Z 393b6012-d7d9-4596-8661-241491e84d11 azul.boto3 s3.HeadObject: Got 404 response |
| 2026-04-24 22:16:07.499 | [INFO] 2026-04-24T22:16:07.499Z 393b6012-d7d9-4596-8661-241491e84d11 azul.boto3 s3.HeadObject: … without a response body |
| 2026-04-24 22:16:07.500 | [INFO] 2026-04-24T22:16:07.500Z 393b6012-d7d9-4596-8661-241491e84d11 azul.terra … with request headers [(&#x27;x-goog-api-client&#x27;, &#x27;cred-type&#x2f;sa&#x27;), (&#x27;authorization&#x27;, &#x27;REDACTED&#x27;)] |
| 2026-04-24 22:16:07.500 | [INFO] 2026-04-24T22:16:07.500Z 393b6012-d7d9-4596-8661-241491e84d11 azul.terra … without a request body |
| 2026-04-24 22:16:07.500 | [INFO] 2026-04-24T22:16:07.500Z 393b6012-d7d9-4596-8661-241491e84d11 azul.terra Making GET request to &#x27;https:&#x2f;&#x2f;jade.datarepo-dev.broadinstitute.org&#x2f;ga4gh&#x2f;drs&#x2f;v1&#x2f;objects&#x2f;v2_4ffefed7-76a2-4807-8f3b-b4dde17ad3b5&#x27; |
| 2026-04-24 22:16:07.993 | [INFO] 2026-04-24T22:16:07.993Z 393b6012-d7d9-4596-8661-241491e84d11 azul.terra … with response headers HTTPHeaderDict({&#x27;Date&#x27;: &#x27;Fri, 24 Apr 2026 22:16:07 GMT&#x27;, &#x27;Server&#x27;: &#x27;Apache&#x27;, &#x27;X-Frame-Options&#x27;: &#x27;SAMEORIGIN&#x27;, &#x27;Access-Control-Allow-Headers&#x27;: &#x27;DNT,User-Agent,X-Requested-With,If-Modified-Since,Cache-Control,Content-Type,Range,Authorization,Accept,Referer,X-App-Id,Origin&#x27;, &#x27;Access-Control-Allow-Methods&#x27;: &#x27;GET,POST,DELETE,PUT,PATCH,OPTIONS,HEAD&#x27;, &#x27;X-Content-Type-Options&#x27;: &#x27;nosniff&#x27;, &#x27;Strict-Transport-Security&#x27;: &#x27;max-age=31536000;includeSubDomains&#x27;, &#x27;Cache-Control&#x27;: &#x27;no-cache,no-store,must-revalidate&#x27;, &#x27;X-Request-ID&#x27;: &#x27;a00kzY21&#x27;, &#x27;Content-Type&#x27;: &#x27;application&#x2f;json&#x27;, &#x27;Content-Length&#x27;: &#x27;2495&#x27;, &#x27;Set-Cookie&#x27;: &#x27;JSESSIONID=1EB5041F637C39BCBD5093E66282B55D; Path=&#x2f;; Secure; HttpOnly; SameSite=Lax&#x27;, &#x27;Vary&#x27;: &#x27;Origin&#x27;, &#x27;Via&#x27;: &#x27;1.1 google&#x27;, &#x27;Alt-Svc&#x27;: &#x27;h3=&quot;:443&quot;; ma=2592000&#x27;}) |
| 2026-04-24 22:16:07.993 | [INFO] 2026-04-24T22:16:07.993Z 393b6012-d7d9-4596-8661-241491e84d11 azul.terra Got 200 response after 0.493s from GET to https:&#x2f;&#x2f;jade.datarepo-dev.broadinstitute.org&#x2f;ga4gh&#x2f;drs&#x2f;v1&#x2f;objects&#x2f;v2_4ffefed7-76a2-4807-8f3b-b4dde17ad3b5 |
| 2026-04-24 22:16:07.993 | [INFO] 2026-04-24T22:16:07.993Z 393b6012-d7d9-4596-8661-241491e84d11 azul.terra … with a response body of length 2495 starting in b&#x27;{&quot;id&quot;:&quot;v2_4ffefed7-76a2-4807-8f3b-b4dde17ad3b5&quot;,&quot;name&quot;:&quot;10137_10137.merged.matefixed.sorted.markeddups.recal.bam&quot;,&quot;self_uri&quot;:&quot;drs:&#x2f;&#x2f;jade.datarepo-dev.broadinstitute.org&#x2f;v2_4ffefed7-76a2-4807-8f3b-b4dde17ad3b5&quot;,&quot;size&quot;:1840,&quot;created_time&quot;:&quot;2023-02-24T20:08:34.007Z&quot;,&quot;updated_time&quot;:&quot;2023-02-24T20:08:34.007Z&quot;,&quot;version&quot;:&quot;0&quot;,&quot;mime_type&quot;:&quot;application&#x2f;octet-stream&quot;,&quot;checksums&quot;:[{&quot;checksum&quot;:&quot;c2c3a933&quot;,&quot;type&quot;:&quot;crc32c&quot;},{&quot;checksum&quot;:&quot;a5a2273ce9a6df96e3a0861a12a8a2ba&quot;,&quot;type&quot;:&quot;md5&quot;}],&quot;access_methods&quot;:[{&quot;type&quot;:&quot;gs&quot;,&quot;access_url&quot;:{&quot;url&quot;:&quot;gs:&#x2f;&#x2f;fc-73435f45-96f6-4077-8dea-89782c5c1de5&#x2f;10137&#x2f;10137.merged.matefixed.sorted.markeddups.recal.bam&quot;,&quot;headers&quot;:null},&quot;access_id&quot;:&quot;gcp-us-central1*74089fe0-c732-40b0-b828-d7e091f0f54b&quot;,&quot;region&quot;:&quot;us-central1&quot;,&quot;authorizations&quot;:{&quot;supported_types&quot;:[&quot;BearerAuth&quot;],&quot;passport_auth_issuers&quot;:null,&quot;bearer_auth_issuers&quot;:null}},{&quot;type&quot;:&quot;https&quot;,&quot;access_url&quot;:{&quot;url&quot;:&quot;https:&#x2f;&#x2f;www.googleapis.com&#x2f;storage&#x2f;v1&#x2f;b&#x2f;fc-73435f45-96f6-4077-8dea-89782c5c1de5&#x2f;o&#x2f;10137%2F10137.merged.matefixed.sorted.markeddups.recal.bam...&#x27; |
| 2026-04-24 22:16:07.993 | [INFO] 2026-04-24T22:16:07.993Z 393b6012-d7d9-4596-8661-241491e84d11 azul.terra … without a request body |
| 2026-04-24 22:16:07.993 | [INFO] 2026-04-24T22:16:07.993Z 393b6012-d7d9-4596-8661-241491e84d11 azul.terra Making GET request to &#x27;https:&#x2f;&#x2f;jade.datarepo-dev.broadinstitute.org&#x2f;ga4gh&#x2f;drs&#x2f;v1&#x2f;objects&#x2f;v2_4ffefed7-76a2-4807-8f3b-b4dde17ad3b5&#x2f;access&#x2f;gcp-us-central1*74089fe0-c732-40b0-b828-d7e091f0f54b&#x27; |
| 2026-04-24 22:16:07.994 | [INFO] 2026-04-24T22:16:07.994Z 393b6012-d7d9-4596-8661-241491e84d11 azul.terra … with request headers [(&#x27;x-goog-api-client&#x27;, &#x27;cred-type&#x2f;sa&#x27;), (&#x27;authorization&#x27;, &#x27;REDACTED&#x27;)] |
| 2026-04-24 22:16:08.662 | [INFO] 2026-04-24T22:16:08.662Z 393b6012-d7d9-4596-8661-241491e84d11 azul.terra … with a response body of length 98 being b&#x27;{&quot;msg&quot;:&quot;Snapshot requires an x-user-project header for DRS access URL requests&quot;,&quot;status_code&quot;:400}&#x27; |
| 2026-04-24 22:16:08.662 | [INFO] 2026-04-24T22:16:08.662Z 393b6012-d7d9-4596-8661-241491e84d11 azul.terra … with response headers HTTPHeaderDict({&#x27;Date&#x27;: &#x27;Fri, 24 Apr 2026 22:16:08 GMT&#x27;, &#x27;Server&#x27;: &#x27;Apache&#x27;, &#x27;X-Frame-Options&#x27;: &#x27;SAMEORIGIN&#x27;, &#x27;Access-Control-Allow-Headers&#x27;: &#x27;DNT,User-Agent,X-Requested-With,If-Modified-Since,Cache-Control,Content-Type,Range,Authorization,Accept,Referer,X-App-Id,Origin&#x27;, &#x27;Access-Control-Allow-Methods&#x27;: &#x27;GET,POST,DELETE,PUT,PATCH,OPTIONS,HEAD&#x27;, &#x27;X-Content-Type-Options&#x27;: &#x27;nosniff&#x27;, &#x27;Strict-Transport-Security&#x27;: &#x27;max-age=31536000;includeSubDomains&#x27;, &#x27;Cache-Control&#x27;: &#x27;no-cache,no-store,must-revalidate&#x27;, &#x27;X-Request-ID&#x27;: &#x27;axr3o84G&#x27;, &#x27;Content-Type&#x27;: &#x27;application&#x2f;json&#x27;, &#x27;Content-Length&#x27;: &#x27;98&#x27;, &#x27;Set-Cookie&#x27;: &#x27;JSESSIONID=85A991CEC23DAA65AEB4AA5000E69046; Path=&#x2f;; Secure; HttpOnly; SameSite=Lax&#x27;, &#x27;Vary&#x27;: &#x27;Origin&#x27;, &#x27;Via&#x27;: &#x27;1.1 google&#x27;, &#x27;Alt-Svc&#x27;: &#x27;h3=&quot;:443&quot;; ma=2592000&#x27;}) |
| 2026-04-24 22:16:08.662 | [INFO] 2026-04-24T22:16:08.662Z 393b6012-d7d9-4596-8661-241491e84d11 azul.terra Got 400 response after 0.668s from GET to https:&#x2f;&#x2f;jade.datarepo-dev.broadinstitute.org&#x2f;ga4gh&#x2f;drs&#x2f;v1&#x2f;objects&#x2f;v2_4ffefed7-76a2-4807-8f3b-b4dde17ad3b5&#x2f;access&#x2f;gcp-us-central1*74089fe0-c732-40b0-b828-d7e091f0f54b |
| 2026-04-24 22:16:08.664 | [ERROR] 2026-04-24T22:16:08.662Z 393b6012-d7d9-4596-8661-241491e84d11 azul-service-anvildev Caught exception for path &#x2f;fetch&#x2f;repository&#x2f;files&#x2f;{file_uuid} Traceback (most recent call last):   File &quot;&#x2f;var&#x2f;task&#x2f;chalice&#x2f;app.py&quot;, line 1922, in _get_view_function_response     response = view_function(**function_args)   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;service&#x2f;repository_controller.py&quot;, line 261, in get_fetch_repository_files     body = self.download_file(file_uuid, fetch=True)   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;service&#x2f;repository_controller.py&quot;, line 313, in download_file     return self._download_file(catalog=catalog,            ~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^                                fetch=fetch,                                ^^^^^^^^^^^^     ...&lt;2 lines&gt;...                                headers=headers,                                ^^^^^^^^^^^^^^^^                                authentication=authentication)                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;service&#x2f;repository_controller.py&quot;, line 417, in _download_file     download.update(authentication)     ~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;plugins&#x2f;repository&#x2f;tdr.py&quot;, line 285, in update     access = drs_client.get(access_method=AccessMethod.gs)   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;drs.py&quot;, line 365, in get     return self._get(access_method)            ~~~~~~~~~^^^^^^^^^^^^^^^   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;drs.py&quot;, line 387, in _get     return self._get_access(access_id, AccessMethod.https)            ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^   File &quot;&#x2f;var&#x2f;task&#x2f;azul&#x2f;drs.py&quot;, line 427, in _get_access     raise DRSStatusException(url, response) azul.drs.DRSStatusException: (&#x27;Unexpected response from https:&#x2f;&#x2f;jade.datarepo-dev.broadinstitute.org&#x2f;ga4gh&#x2f;drs&#x2f;v1&#x2f;objects&#x2f;v2_4ffefed7-76a2-4807-8f3b-b4dde17ad3b5&#x2f;access&#x2f;gcp-us-central1*74089fe0-c732-40b0-b828-d7e091f0f54b&#x27;, 400, b&#x27;{&quot;msg&quot;:&quot;Snapshot requires an x-user-project header for DRS access URL requests&quot;,&quot;status_code&quot;:400}&#x27;) |
| 2026-04-24 22:16:08.664 | [INFO] 2026-04-24T22:16:08.664Z 393b6012-d7d9-4596-8661-241491e84d11 azul.chalice … with a response body of length 133 being {&quot;Code&quot;: &quot;InternalServerError&quot;, &quot;Message&quot;: &quot;An internal server error occurred.&quot;, &quot;RequestId&quot;: &quot;393b6012-d7d9-4596-8661-241491e84d11&quot;} |
| 2026-04-24 22:16:08.664 | [INFO] 2026-04-24T22:16:08.664Z 393b6012-d7d9-4596-8661-241491e84d11 azul.chalice Returning 500 response with headers {&quot;headers&quot;: {&quot;Access-Control-Allow-Origin&quot;: &quot;*&quot;, &quot;Access-Control-Allow-Headers&quot;: &quot;Authorization,Content-Type,X-Amz-Date,X-Amz-Security-Token,X-Api-Key&quot;, &quot;Content-Security-Policy&quot;: &quot;default-src &#x27;self&#x27;;img-src &#x27;self&#x27; data:;script-src &#x27;self&#x27;;style-src &#x27;self&#x27;;frame-ancestors &#x27;none&#x27;;form-action &#x27;self&#x27;&quot;, &quot;Referrer-Policy&quot;: &quot;strict-origin-when-cross-origin&quot;, &quot;Strict-Transport-Security&quot;: &quot;max-age=63072000; includeSubDomains; preload&quot;, &quot;X-Content-Type-Options&quot;: &quot;nosniff&quot;, &quot;X-Frame-Options&quot;: &quot;DENY&quot;, &quot;X-XSS-Protection&quot;: &quot;1; mode=block&quot;, &quot;Cache-Control&quot;: &quot;no-store&quot;}}. |
| 2026-04-24 22:16:08.666 | REPORT RequestId: 393b6012-d7d9-4596-8661-241491e84d11 Duration: 1194.81 ms Billed Duration: 1195 ms Memory Size: 768 MB Max Memory Used: 183 MB |
| 2026-04-24 22:16:08.666 | END RequestId: 393b6012-d7d9-4596-8661-241491e84d11 |
---

--- a/README.md
+++ b/README.md
@@ -676,6 +676,67 @@ These steps are performed once per deployment (multiple times per project).
 
     Follow the prompts to store the client secret.
 
+### 3.2.3 Terra workspace for requester-pays with TDR
+
+TDR supports requester-pays via Google Cloud to allow for data egress charges
+(e.g. from mirroring) to be billed to a specific [Terra billing project] instead
+of the bucket owner. Billing projects are a concept internal to Terra, related
+to but distinct from Google Cloud billing accounts.
+
+[Terra billing project]: https://support.terra.bio/hc/en-us/articles/360026182251-How-to-set-up-billing-in-Terra-GCP
+
+To use requester-pays, Azul must be configured with a Google Project ID
+associated with a Terra workspace that belongs to the chosen Terra billing
+project. This workspace must be created and configured manually. To set up a
+workspace for requester-pays in a given deployment:
+
+1. If the deployment's service accounts have not already been
+   [registered with SAM](#234-google-cloud-tdr-and-sam), complete the steps in
+   that section before proceeding further.
+
+2. Go to [Terra](https://app.terra.bio/#workspace) and create a new workspace.
+   Give it the name `"azul-requester_pays-${AZUL_DEPLOYMENT_STAGE}"`. This step
+   is usually skipped for personal deployments, which share a workspace with
+   their associated sandbox deployment.
+
+3. Select the Terra billing project to be used for requester-pays.
+
+4. Add the deployment's public service account as a collaborator and assign it
+   the role of "Writer". The service account's email follows the pattern
+   `"${AZUL_GOOGLE_SERVICE_ACCOUNT_PUBLIC}@${GOOGLE_PROJECT}.iam.gserviceaccount.com"`
+
+5. If the deployment will be used to mirror managed-access files, repeat step 4
+   for the indexer service account.
+
+In Terra's dev environment, snapshots' associated storage buckets are never
+public, even if the associated snapshot is public. This means that a
+deployment's public service account must be explicitly granted access to the
+public snapshots before it can be used for requester-pays.
+
+This access is managed via a group analogous to the group that grants the
+indexer service account access to all indexed snapshots. The public group is
+set up once per main deployment in Terra dev. Create a new Terra group and ask
+a representative from the Terra team to grant the group access to each public
+snapshot indexed by Azul.
+
+Once the public group is set up:
+
+1. Ask the system administrator to add your deployment's public service account
+   to the group.
+
+2. Verify that the public service account cannot access any non-public
+   snapshots.
+
+Then, to enable requester-pays for that deployment:
+
+1. Copy the workspace's Google Project ID from the "Cloud Information" side
+   panel.
+
+2. Set `"AZUL_TDR_REQUESTER_PAYS_PROJECT"` to the copied project ID in the
+   chosen deployment's `environment.py`.
+
+3. `_refresh` and `make deploy`.
+
 ## 3.3 Provisioning cloud infrastructure
 
 Once you've configured the project and your personal deployment or a shared

--- a/UPGRADING.rst
+++ b/UPGRADING.rst
@@ -19,6 +19,30 @@ branch that does not have the listed changes, the steps would need to be
 reverted. This is all fairly informal and loosely defined. Hopefully we won't
 have too many entries in this file.
 
+#7794 Use requester-pays for mirroring in AnVIL deployments
+===========================================================
+
+For each of your personal deployments colocated with ``anvildev``:
+
+1. Ask the system administrator to add both your deployment's public service
+   account and indexer service account to the ``azul-requester_pays-anvilbox``
+   Terra workspace and assign them the role of "Writer". You can find the
+   service accounts' emails by running ::
+
+    # public service account
+    python -c 'from azul.terra import TDRClient; print(TDRClient.for_anonymous_user().credentials.service_account_email)'
+
+    # indexer service account
+    python -c 'from azul.terra import TDRClient; print(TDRClient.for_indexer().credentials.service_account_email)'
+
+2. Ask the system administrator to add your deployment's public service account
+   to the ``azul-anvil-public-dev`` Terra group.
+
+3. Set the ``AZUL_TDR_REQUESTER_PAYS_PROJECT`` variable in ``environment.py``
+   to the Google Project ID of the Terra workspace ``azul-requester_pays-anvilbox``.
+   As always, use the sandbox deployment's ``environment.py`` as a model when
+   upgrading personal deployments.
+
 
 #7993 Tighten verification of access tokens
 ===========================================

--- a/deployments/anvilbox/environment.py
+++ b/deployments/anvilbox/environment.py
@@ -147,6 +147,7 @@ def env() -> Mapping[str, str | None]:
         'AZUL_SAM_SERVICE_URL': 'https://sam.dsde-dev.broadinstitute.org',
         'AZUL_DUOS_SERVICE_URL': 'https://consent.dsde-dev.broadinstitute.org',
         'AZUL_TERRA_SERVICE_URL': 'https://firecloud-orchestration.dsde-dev.broadinstitute.org',
+        'AZUL_TDR_REQUESTER_PAYS_PROJECT': 'terra-dev-8c6454ca',
         'azul_ecm_service_url': 'https://externalcreds.dsde-dev.broadinstitute.org',
 
         **(

--- a/deployments/anvilbox/environment.py
+++ b/deployments/anvilbox/environment.py
@@ -184,9 +184,4 @@ def env() -> Mapping[str, str | None]:
         'AZUL_GOOGLE_OAUTH2_CLIENT_ID': '561542988117-cpo2avhomdh6t7fetp91js78cdhm9p47.apps.googleusercontent.com',
 
         'AZUL_ENABLE_MIRRORING': '1',
-
-        # FIXME: Revert, once the underlying issue with requester-pays is fixed
-        #        https://github.com/DataBiosphere/azul/issues/7955
-        #
-        'azul_it_flags': 'no_mirror',
     }

--- a/deployments/anvildev/environment.py
+++ b/deployments/anvildev/environment.py
@@ -164,9 +164,4 @@ def env() -> Mapping[str, str | None]:
             'workspace_id': 'T09P9H91S',  # ucsc-gi.slack.com
             'channel_id': 'C04K4BQET7G'  # #team-boardwalk-anvildev
         }),
-
-        # FIXME: Revert, once the underlying issue with requester-pays is fixed
-        #        https://github.com/DataBiosphere/azul/issues/7955
-        #
-        'azul_it_flags': 'no_mirror',
     }

--- a/deployments/anvildev/environment.py
+++ b/deployments/anvildev/environment.py
@@ -134,6 +134,7 @@ def env() -> Mapping[str, str | None]:
         'AZUL_SAM_SERVICE_URL': 'https://sam.dsde-dev.broadinstitute.org',
         'AZUL_DUOS_SERVICE_URL': 'https://consent.dsde-dev.broadinstitute.org',
         'AZUL_TERRA_SERVICE_URL': 'https://firecloud-orchestration.dsde-dev.broadinstitute.org',
+        'AZUL_TDR_REQUESTER_PAYS_PROJECT': 'terra-dev-789f8dd1',
         'azul_ecm_service_url': 'https://externalcreds.dsde-dev.broadinstitute.org',
 
         'AZUL_ENABLE_MONITORING': '1',

--- a/deployments/anvilprod/environment.py
+++ b/deployments/anvilprod/environment.py
@@ -1393,6 +1393,7 @@ def env() -> Mapping[str, str | None]:
         'AZUL_SAM_SERVICE_URL': 'https://sam.dsde-prod.broadinstitute.org',
         'AZUL_DUOS_SERVICE_URL': 'https://consent.dsde-prod.broadinstitute.org',
         'AZUL_TERRA_SERVICE_URL': 'https://firecloud-orchestration.dsde-prod.broadinstitute.org',
+        'AZUL_TDR_REQUESTER_PAYS_PROJECT': 'terra-20cf0e48',
         'azul_ecm_service_url': 'https://externalcreds.dsde-prod.broadinstitute.org',
 
         'AZUL_ENABLE_MONITORING': '1',

--- a/deployments/anvilprod/environment.py
+++ b/deployments/anvilprod/environment.py
@@ -1430,10 +1430,5 @@ def env() -> Mapping[str, str | None]:
 
         'AZUL_MIRROR_BUCKET': 'anvilproject',
 
-        'AZUL_MIRRORING_CONCURRENCY': '128',
-
-        # FIXME: Revert, once the underlying issue with requester-pays is fixed
-        #        https://github.com/DataBiosphere/azul/issues/7955
-        #
-        'azul_it_flags': 'no_mirror',
+        'AZUL_MIRRORING_CONCURRENCY': '128'
     }

--- a/deployments/dev/environment.py
+++ b/deployments/dev/environment.py
@@ -127,9 +127,7 @@ dcp2_sources = union({}, 105, delta([
     source('bigquery', 'datarepo-dev-59d37b9a', 'hca_dev_946c5add47d1402a97bba5af97e8bce7__20210831_20210903'),
     source('bigquery', 'datarepo-dev-788c3b52', 'hca_dev_955dfc2ca8c64d04aa4d907610545d11__20210831_20210903', no_mirror),  # noqa: E501, all files above mirror size limit
     source('bigquery', 'datarepo-dev-4b88b45b', 'hca_dev_962bd805eb894c54bad2008e497d1307__20210830_20210903'),
-    # FIXME: Make this no_ma_mirror again, once the underlying issue with requester-pays is fixed
-    #        https://github.com/DataBiosphere/azul/issues/7955
-    source('bigquery', 'datarepo-dev-02c59b72', 'hca_dev_99101928d9b14aafb759e97958ac7403__20210830_20210903'),
+    source('bigquery', 'datarepo-dev-02c59b72', 'hca_dev_99101928d9b14aafb759e97958ac7403__20210830_20210903', no_ma_mirror),  # noqa: E501
     source('bigquery', 'datarepo-dev-a6312a94', 'hca_dev_992aad5e7fab46d9a47ddf715e8cfd24__20210830_20210903'),
     source('bigquery', 'datarepo-dev-75589244', 'hca_dev_996120f9e84f409fa01e732ab58ca8b9__20210827_20210903'),
     source('bigquery', 'datarepo-dev-d4b988d6', 'hca_dev_a004b1501c364af69bbd070c06dbc17d__20210830_20210903', no_ma_mirror),  # noqa: E501

--- a/deployments/hammerbox/environment.py
+++ b/deployments/hammerbox/environment.py
@@ -1445,9 +1445,4 @@ def env() -> Mapping[str, str | None]:
         'AZUL_ENABLE_VERBATIM_RELATIONS': '0',
 
         'AZUL_ENABLE_MIRRORING': '1',
-
-        # FIXME: Revert, once the underlying issue with requester-pays is fixed
-        #        https://github.com/DataBiosphere/azul/issues/7955
-        #
-        'azul_it_flags': 'no_mirror',
     }

--- a/deployments/hammerbox/environment.py
+++ b/deployments/hammerbox/environment.py
@@ -1407,6 +1407,7 @@ def env() -> Mapping[str, str | None]:
         'AZUL_SAM_SERVICE_URL': 'https://sam.dsde-prod.broadinstitute.org',
         'AZUL_DUOS_SERVICE_URL': 'https://consent.dsde-prod.broadinstitute.org',
         'AZUL_TERRA_SERVICE_URL': 'https://firecloud-orchestration.dsde-prod.broadinstitute.org',
+        'AZUL_TDR_REQUESTER_PAYS_PROJECT': 'terra-fb27b873',
         'azul_ecm_service_url': 'https://externalcreds.dsde-prod.broadinstitute.org',
 
         # Personal deployments & `hammerbox` share an ES domain with `anvilprod`

--- a/deployments/sandbox/environment.py
+++ b/deployments/sandbox/environment.py
@@ -129,9 +129,7 @@ dcp2_sources = union({}, 105, delta([
     source('bigquery', 'datarepo-dev-59d37b9a', 'hca_dev_946c5add47d1402a97bba5af97e8bce7__20210831_20210903'),
     source('bigquery', 'datarepo-dev-788c3b52', 'hca_dev_955dfc2ca8c64d04aa4d907610545d11__20210831_20210903', no_mirror),  # noqa: E501, all files above mirror size limit
     source('bigquery', 'datarepo-dev-4b88b45b', 'hca_dev_962bd805eb894c54bad2008e497d1307__20210830_20210903'),
-    # FIXME: Make this no_ma_mirror again, once the underlying issue with requester-pays is fixed
-    #        https://github.com/DataBiosphere/azul/issues/7955
-    source('bigquery', 'datarepo-dev-02c59b72', 'hca_dev_99101928d9b14aafb759e97958ac7403__20210830_20210903'),
+    source('bigquery', 'datarepo-dev-02c59b72', 'hca_dev_99101928d9b14aafb759e97958ac7403__20210830_20210903', no_ma_mirror),  # noqa: E501
     source('bigquery', 'datarepo-dev-a6312a94', 'hca_dev_992aad5e7fab46d9a47ddf715e8cfd24__20210830_20210903'),
     source('bigquery', 'datarepo-dev-75589244', 'hca_dev_996120f9e84f409fa01e732ab58ca8b9__20210827_20210903'),
     source('bigquery', 'datarepo-dev-d4b988d6', 'hca_dev_a004b1501c364af69bbd070c06dbc17d__20210830_20210903', no_ma_mirror),  # noqa: E501

--- a/deployments/tempdev/environment.py
+++ b/deployments/tempdev/environment.py
@@ -162,9 +162,4 @@ def env() -> Mapping[str, str | None]:
         'AZUL_GOOGLE_OAUTH2_CLIENT_ID': '807674395527-erth0gf1m7qme5pe6bu384vpdfjh06dg.apps.googleusercontent.com',
 
         'AZUL_ENABLE_MIRRORING': '1',
-
-        # FIXME: Revert, once the underlying issue with requester-pays is fixed
-        #        https://github.com/DataBiosphere/azul/issues/7955
-        #
-        'azul_it_flags': 'no_mirror',
     }

--- a/environment.py
+++ b/environment.py
@@ -762,6 +762,14 @@ def env() -> Mapping[str, str | None]:
         #
         'AZUL_TERRA_SERVICE_URL': None,
 
+        # The GCP project ID of the Terra workspace to charge for file downloads
+        # from TDR while mirroring. If left unset, egress charges are incurred
+        # to the owner of the GCS bucket the files are stored in. Otherwise, the
+        # egress will be charged to the GCP billing account associated with the
+        # workspace. See section 3.2.3 of the README.
+        #
+        'AZUL_TDR_REQUESTER_PAYS_PROJECT': None,
+
         # OAuth2 Client ID to be used for authenticating users. See section
         # 3.2 of the README
         #

--- a/src/azul/__init__.py
+++ b/src/azul/__init__.py
@@ -355,6 +355,10 @@ class Config:
         return mutable_furl(self.environ['azul_ecm_service_url'])
 
     @property
+    def tdr_requester_pays_project(self) -> str | None:
+        return self.environ.get('AZUL_TDR_REQUESTER_PAYS_PROJECT')
+
+    @property
     def dss_query_prefix(self) -> str:
         return self.environ.get('AZUL_DSS_QUERY_PREFIX', '')
 

--- a/src/azul/drs.py
+++ b/src/azul/drs.py
@@ -356,15 +356,30 @@ class DRSObject:
     _http_client: HttpClient
     _url: furl
 
-    def get(self, access_method: AccessMethod = AccessMethod.https) -> Access:
+    def get(self,
+            access_method: AccessMethod = AccessMethod.https,
+            headers: Mapping[str, str] | None = None
+            ) -> Access:
         """
         Returns access to the content of the data object identified by the
-        given URI. The scheme of the URL in the returned access object depends
-        on the access method specified.
-        """
-        return self._get(access_method)
+        given URI.
 
-    def _get(self, access_method: AccessMethod) -> Access:
+        :param access_method: The type of access method to use from the object
+                              response. The scheme of the URL in the returned
+                              access object depends on the access method
+                              specified.
+
+        :param headers: Optional request headers for accessing the object. Note
+                        that this argument is only applied for requests to
+                        .../objects/{object_id}/access/{access_id}/, and not to
+                        .../objects/{object_id}.
+        """
+        return self._get(access_method, headers)
+
+    def _get(self,
+             access_method: AccessMethod,
+             headers: Mapping[str, str] | None
+             ) -> Access:
         url = self._url
         while True:
             response = self._request(url)
@@ -384,9 +399,9 @@ class DRSObject:
                     # https://github.com/ga4gh/data-repository-service-schemas/issues/361
                     assert access_method is AccessMethod.gs, R(
                         'Unexpected access method', access_method)
-                    return self._get_access(access_id, AccessMethod.https)
+                    return self._get_access(access_id, AccessMethod.https, headers)
                 elif access_id is not None:
-                    return self._get_access(access_id, access_method)
+                    return self._get_access(access_id, access_method, headers)
                 elif access_url is not None:
                     scheme = furl(access_url['url']).scheme
                     assert scheme == access_method.scheme, R(
@@ -403,11 +418,15 @@ class DRSObject:
             else:
                 raise DRSStatusException(url, response)
 
-    def _get_access(self, access_id: str, access_method: AccessMethod) -> Access:
+    def _get_access(self,
+                    access_id: str,
+                    access_method: AccessMethod,
+                    headers: Mapping[str, str] | None
+                    ) -> Access:
         url = self._url.copy()
         url.path.add(['access', access_id])
         while True:
-            response = self._request(url)
+            response = self._request(url, headers=headers)
             if response.status == 200:
                 response_data = json_dict(json.loads(response.data))
                 scheme = furl(json_str(response_data['url'])).scheme
@@ -429,8 +448,8 @@ class DRSObject:
             else:
                 raise DRSStatusException(url, response)
 
-    def _request(self, url: furl) -> urllib3.BaseHTTPResponse:
-        return self._http_client.request('GET', str(url), redirect=False)
+    def _request(self, url: furl, **kwargs) -> urllib3.BaseHTTPResponse:
+        return self._http_client.request('GET', str(url), **kwargs, redirect=False)
 
 
 class DRSStatusException(Exception):

--- a/src/azul/drs.py
+++ b/src/azul/drs.py
@@ -414,12 +414,13 @@ class DRSObject:
                 assert scheme == access_method.scheme, R(
                     'Unexpected access URL scheme', scheme)
                 access_url = json_str(response_data['url'])
-                headers = response_data.get('headers')
-                if headers is None:
-                    access_headers = None
-                else:
-                    access_headers = {k: json_str(v) for k, v in json_dict(headers).items()}
-                return Access(method=access_method, url=access_url, headers=access_headers)
+                access_headers = response_data.get('headers')
+                return Access(method=access_method,
+                              url=access_url,
+                              headers=None if access_headers is None else {
+                                  k: json_str(v)
+                                  for k, v in json_dict(access_headers).items()
+                              })
             elif response.status == 202:
                 wait_time = int(response.headers['retry-after'])
                 time.sleep(wait_time)

--- a/src/azul/indexer/mirror_service.py
+++ b/src/azul/indexer/mirror_service.py
@@ -42,9 +42,6 @@ from azul.auth import (
 from azul.deployment import (
     aws,
 )
-from azul.drs import (
-    AccessMethod,
-)
 from azul.http import (
     HasCachedHttpClient,
 )
@@ -77,6 +74,9 @@ from azul.plugins import (
     File,
     RepositoryFileDownload,
     RepositoryPlugin,
+)
+from azul.plugins.repository.tdr import (
+    TDRFileDownload,
 )
 from azul.queues import (
     Action,
@@ -788,7 +788,7 @@ class MirrorWorkerService(MirrorService, HasCachedHttpClient):
                            content_type='application/json',
                            overwrite=False)
 
-    def _repository_url(self, file: File) -> furl:
+    def _repository_url(self, file: File) -> str:
         assert config.is_tdr_enabled(self.catalog), R(
             'Only TDR catalogs are supported', self.catalog)
         assert file.drs_uri is not None, R(
@@ -799,10 +799,14 @@ class MirrorWorkerService(MirrorService, HasCachedHttpClient):
             authentication = None
         else:
             authentication = indexer_authentication
-        object = self.repository_plugin.drs_object(file.drs_uri, authentication)
-        access = object.get(AccessMethod.gs)
-        assert access.method is AccessMethod.https, access
-        return furl(access.url)
+        download = TDRFileDownload(plugin=self.repository_plugin,
+                                   file=file,
+                                   replica=None,
+                                   token=None)
+        download.update(authentication)
+        assert download.retry_after is None
+        assert download.location is not None
+        return download.location
 
     def _download(self, file: File, part: FilePart | None = None) -> bytes:
         url = self._repository_url(file)
@@ -817,7 +821,7 @@ class MirrorWorkerService(MirrorService, HasCachedHttpClient):
             expected_status = 206
         # Ideally we would stream the response, but boto only supports uploading
         # from streams that are seekable.
-        response = self._http_client.request('GET', str(url), headers=headers)
+        response = self._http_client.request('GET', url, headers=headers)
         if response.status == expected_status:
             actual_size = len(response.data)
             log.info('Downloaded %d bytes in %.3fs from file %r',

--- a/src/azul/indexer/mirror_service.py
+++ b/src/azul/indexer/mirror_service.py
@@ -802,7 +802,8 @@ class MirrorWorkerService(MirrorService, HasCachedHttpClient):
         download = TDRFileDownload(plugin=self.repository_plugin,
                                    file=file,
                                    replica=None,
-                                   token=None)
+                                   token=None,
+                                   requester_pays=True)
         download.update(authentication)
         assert download.retry_after is None
         assert download.location is not None

--- a/src/azul/plugins/repository/tdr.py
+++ b/src/azul/plugins/repository/tdr.py
@@ -14,6 +14,7 @@ from typing import (
     TypeVar,
 )
 
+import attrs
 from chalice import (
     UnauthorizedError,
 )
@@ -270,7 +271,9 @@ class TDRPlugin[TDR_BUNDLE: TDRBundle,
                     yield match
 
 
+@attrs.define(auto_attribs=True, kw_only=True)
 class TDRFileDownload(RepositoryFileDownload):
+    requester_pays: bool = False
     _location: str | None = None
 
     def update(self, authentication: Authentication | None) -> None:
@@ -280,8 +283,12 @@ class TDRFileDownload(RepositoryFileDownload):
             assert self.location is None, self
             assert self.retry_after is None, self
         else:
+            if self.requester_pays and config.tdr_requester_pays_project is not None:
+                headers = {'x-user-project': config.tdr_requester_pays_project}
+            else:
+                headers = None
             drs_client = self._plugin.drs_object(self.file.drs_uri, authentication)
-            access = drs_client.get(access_method=AccessMethod.gs)
+            access = drs_client.get(access_method=AccessMethod.gs, headers=headers)
             assert access.method is AccessMethod.https, R(str(access.method))
             assert access.headers is None, R(str(access.headers))
             signed_url = access.url

--- a/test/integration_test.py
+++ b/test/integration_test.py
@@ -550,10 +550,7 @@ class IndexingIntegrationTest(IntegrationTestCase):
         for catalog in catalogs:
             self._test_manifest(catalog.name)
             self._test_manifest_tagging_race(catalog.name)
-            # FIXME: Revert, once the underlying issue with requester-pays is fixed
-            #        https://github.com/DataBiosphere/azul/issues/7955
-            if mirror:
-                self._test_curl_manifest(catalog.name)
+            self._test_curl_manifest(catalog.name)
             self._test_drs(catalog.name)
             self._test_repository_files(catalog.name)
             self._test_managed_access(catalog=catalog.name,

--- a/test/integration_test.py
+++ b/test/integration_test.py
@@ -515,11 +515,7 @@ class IndexingIntegrationTest(IntegrationTestCase):
                     # If test_mirroring is run for the catalog, ensure that the
                     # source is not flagged as no_mirror so that we can test
                     # downloading a mirrored file
-                    #
-                    # FIXME: Revert, once the underlying issue with requester-pays is fixed
-                    #        https://github.com/DataBiosphere/azul/issues/7955
-                    #
-                    mirror=True and self._mirror_service(catalog.name).may_mirror()
+                    mirror=mirror and self._mirror_service(catalog.name).may_mirror()
                 ).ref
                 ma_source = self._select_source(catalog.name, public=False)
                 ma_source = None if ma_source is None else ma_source.ref

--- a/test/integration_test.py
+++ b/test/integration_test.py
@@ -550,7 +550,10 @@ class IndexingIntegrationTest(IntegrationTestCase):
         for catalog in catalogs:
             self._test_manifest(catalog.name)
             self._test_manifest_tagging_race(catalog.name)
-            self._test_curl_manifest(catalog.name)
+            # FIXME: Re-enable curl manifest IT for AnVIL
+            #        https://github.com/DataBiosphere/azul/issues/8095
+            if config.tdr_requester_pays_project is None:
+                self._test_curl_manifest(catalog.name)
             self._test_drs(catalog.name)
             self._test_repository_files(catalog.name)
             self._test_managed_access(catalog=catalog.name,

--- a/test/integration_test.py
+++ b/test/integration_test.py
@@ -1364,12 +1364,22 @@ class IndexingIntegrationTest(IntegrationTestCase):
             self.assertNotIn(str(config.tdr_service_url), msg)
             return None
         elif response.status == 403:
-            # FIXME: Treat this as an error if requester-pays is enabled
-            #        https://github.com/DataBiosphere/azul/issues/7794
             msg = json.loads(response.data)['Message']
             prefix = 'DRS server requires requester-pays for '
             self.assertEqual(prefix, msg[:len(prefix)])
-            return None
+            if config.tdr_requester_pays_project is None:
+                # A requester-pays project ought to be configured in every
+                # deployment that reads from a TDR instance where requester-pays
+                # is required.
+                self.fail(msg)
+            else:
+                # We intentionally omit the x-user-project header from all file
+                # download requests except for those made during mirroring.
+                # Consequently, a 403 response is expected when attempting to
+                # download non-mirrored files. This test runs before the
+                # mirroring subtest, so the file being downloaded is never
+                # mirrored.
+                return None
         else:
             self.assertEqual(200, response.status)
             response = json.loads(response.data)

